{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3adceb3",
   "metadata": {},
   "source": [
    "## MAPS analysis\n",
    "\n",
    "this method was refered to Lee et.al,Integrative analysis reveals RNA G-quadruplexes in UTRs are selectively constrained and enriched for functional associations,Nat Commun,2020.\n",
    "\n",
    "#### data prepared \n",
    "the methylation_hg38.RData and mutation_rate_methylation_bins.txt were download from https://zenodo.org/record/4536050/files/methylation_hg38.RData?download=1, this data was used to control methylation levels(Lee, et al. Disrupting upstream translation in mRNAs is associated with human disease. *Nat Communications* 2021)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3c11a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## maps_noG4_g4_data_get.py\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import os,sys\n",
    "wdpath = \"/disk2/G4_MS/frequency/MAPS/data\"\n",
    "gruns = pr.PyRanges(pd.read_table(\"/disk2/G4_MS/reference/genomic_Gruns.bed\", header=None, names=[\"Chromosome\", \"Start\", \"End\" ,\"Strand\", \"grun\"]))\n",
    "promoter = pr.PyRanges(pd.read_table(\"/disk2/G4_MS/promoter_select/promoter_info/promoter.annotion.tsv\"))\n",
    "epromoter = pr.PyRanges(pd.read_table(\"/disk2/G4_MS/promoter_select/promoter_info/promoter_info.annotion.H3K4me3.tsv\"))\n",
    "promoter_pG4 = pr.PyRanges(pd.read_table(\"/disk2/G4_MS/promoter_select/promoter_info/promoter_pG4.info.tsv\").loc[:,[\"Chromosome\", \"Start\",\"End\",\"Strand\",\"Name\",\"gene_id\",\"transcript_id\",\"Score\", \"Strand_b\"]])\n",
    "promoter_eG4 = pr.PyRanges(pd.read_table(\"/disk2/G4_MS/promoter_select/promoter_info/promoter_pG4.info.H3K4me3.tsv\").loc[:,[\"Chromosome\", \"Start\",\"End\",\"Strand\",\"Name\", \"gene_id\",\"transcript_id\",\"Score\", \"Strand_b\"]])\n",
    "all_pG4 = pr.read_bed(\"/disk2/G4_MS/G4_identify/Quadron/hs38d2.quadron.bed\")\n",
    "def gruns_select(gruns, all_pG4, promoter, promoter_pG4, eG4=False):\n",
    "    g4 = \"pG4\"\n",
    "    if eG4:\n",
    "        g4 = \"eG4\"\n",
    "    promoter_noG4 = promoter.subtract(all_pG4,strandedness=False)\n",
    "    promoter_pG4_gruns = promoter_pG4.intersect(gruns, strandedness=\"same\").drop_duplicate_positions().as_df().loc[:,[\"Chromosome\", \"Start\",\"End\",\"transcript_id\",\"gene_id\",\"Strand\"]]\n",
    "    promoter_pG4_gruns.to_csv(os.path.join(wdpath,\"promoter_1000_\"+g4+\".gruns.bed\"),index=False, header=False, sep=\"\\t\")\n",
    "    \n",
    "    promoter_nonG4_gruns = promoter_noG4.intersect(gruns,strandedness=False).drop_duplicate_positions().as_df().loc[:,[\"Chromosome\", \"Start\",\"End\",\"transcript_id\",\"gene_id\",\"Strand\"]]\n",
    "    promoter_nonG4_gruns.to_csv(os.path.join(wdpath, \"protein_promoter_1000_noG4.\"+ g4 + \".gruns.bed\"), index=False, header=False, sep=\"\\t\")\n",
    "    \n",
    "    pre_promoter_pG4 = promoter_pG4.drop_duplicate_positions().as_df().loc[:,[\"Chromosome\", \"Start\",\"End\",\"Name\", \"Strand\",\"transcript_id\",\"gene_id\",\"Strand_b\"]]\n",
    "    if eG4:\n",
    "        pre_promoter_pG4.to_csv(os.path.join(wdpath, \"promoter_pG4.info.H3K4me3.tsv\"),index=False, header=False, sep=\"\\t\")\n",
    "    else:\n",
    "        pre_promoter_pG4.to_csv(os.path.join(wdpath, \"promoter_\"+g4+\".info.tsv\"),index=False, header=False, sep=\"\\t\")\n",
    "    \n",
    "    promoter_stable = promoter_pG4[promoter_pG4.Score >19]\n",
    "    promoter_unstable = promoter_pG4[promoter_pG4.Score <= 19]\n",
    "    \n",
    "    promoter_stable_gruns = promoter_stable.intersect(gruns,strandedness=\"same\").drop_duplicate_positions().as_df().loc[:,[\"Chromosome\", \"Start\",\"End\",\"transcript_id\",\"gene_id\",\"Strand\"]]\n",
    "    promoter_stable_gruns.to_csv(os.path.join(wdpath,\"promoter_1000_stable_\"+g4+\".gruns.bed\"),index=False, header=False, sep=\"\\t\")\n",
    "    \n",
    "    promoter_unstable_gruns = promoter_unstable.intersect(gruns,strandedness=\"same\").drop_duplicate_positions().as_df().loc[:,[\"Chromosome\", \"Start\",\"End\",\"transcript_id\",\"gene_id\",\"Strand\"]]\n",
    "    promoter_unstable_gruns.to_csv(os.path.join(wdpath,\"promoter_1000_unstable_\"+g4+\".gruns.bed\"),index=False, header=False, sep=\"\\t\")\n",
    "    \n",
    "    promoter = promoter.drop_duplicate_positions().as_df().loc[:,[\"Chromosome\",\"Start\", \"End\",\"Promoter_id\", \"gene_id\", \"Strand\",\"transcript_id\"]]\n",
    "    if eG4:\n",
    "        promoter.to_csv(os.path.join(wdpath, \"promoter_info.annotion.H3K4me3.tsv\"),index=False, header=False, sep=\"\\t\")\n",
    "    else:\n",
    "        promoter.to_csv(os.path.join(wdpath, \"promoter.annotion.tsv\"),index=False, header=False, sep=\"\\t\")\n",
    "\n",
    "gruns_select(gruns, all_pG4, promoter, promoter_pG4)\n",
    "gruns_select(gruns, all_pG4, epromoter, promoter_eG4, eG4=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d30f6f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# maps_synonynous.py\n",
    "import os,sys,gzip\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "wdpath = \"/disk2/G4_MS/frequency/MAPS/data\"\n",
    "# refpath = \"/usr/refence/hg38/hg38.fa\"\n",
    "with gzip.open(os.path.join(wdpath,\"cds.gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\"),'rt')as fr,open(os.path.join(wdpath,\"gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv\"),'w')as fw, open(os.path.join(wdpath,\"gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv\"),'w')as fw2:\n",
    "    for line in fr:\n",
    "        uline = line.strip().split(\"\\t\")\n",
    "        chromsome,start,end,ref,alter,ac,an,af,info=uline\n",
    "        if \"synonymous_variant\" in info and \"missense_variant\" not in info:\n",
    "            fw.write(\"\\t\".join([chromsome,start,ref,alter,an,ac,af,\"snv\",'.',info])+\"\\n\")\n",
    "        if \"missense_variant\" in info:\n",
    "            fw2.write(\"\\t\".join([chromsome,start,ref,alter,an,ac,af,\"snv\",'.',info])+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53cbec0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snp_trans.py \n",
    "import sys,os\n",
    "import gzip\n",
    "wdpath = \"/disk2/G4_MS/frequency/MAPS/data\"\n",
    "with gzip.open(os.path.join(wdpath,\"promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\"),'rt')as fr,open(os.path.join(wdpath,\"promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv\"),'w')as fw:\n",
    "    #fw.write(\"chromosome\\tstart\\tref\\talt\\tan\\tac\\taf\\n\")\n",
    "    for line in fr:\n",
    "        uline = line.strip().split(\"\\t\")\n",
    "        chromsome,start,end,ref,alter,ac,an,af,info=uline\n",
    "        fw.write(\"\\t\".join([chromsome,start,ref,alter,an,ac,af])+\"\\n\")\n",
    "        \n",
    "with gzip.open(os.path.join(wdpath,\"active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\"),'rt')as fr,open(os.path.join(wdpath,\"active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv\"),'w')as fw:\n",
    "    #fw.write(\"chromosome\\tstart\\tref\\talt\\tan\\tac\\taf\\n\")\n",
    "    for line in fr:\n",
    "        uline = line.strip().split(\"\\t\")\n",
    "        chromsome,start,end,ref,alter,ac,an,af,info=uline\n",
    "        fw.write(\"\\t\".join([chromsome,start,ref,alter,an,ac,af])+\"\\n\")        \n",
    "\n",
    "with gzip.open(\"/disk2/G4_MS/reference/gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\",'rt')as fr,open(os.path.join(wdpath,\"gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv\"),'w')as fw:\n",
    "    for line in fr:\n",
    "        uline = line.strip().split(\"\\t\")\n",
    "        chromsome,start,end,ref,alter,ac,an,af,info=uline\n",
    "        if \"HIGH\" in info:\n",
    "            fw.write(\"\\t\".join([chromsome,start,ref,alter,an,ac,af,\"snv\",'.',info])+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "177f060d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Title     : TODO\n",
    "# Objective : TODO\n",
    "# Created by: guang\n",
    "# Created on: 2021/9/30\n",
    "# Filname maps.model.R\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "load(\"/disk2/G4_MS/frequency/MAPS/data/methylation_hg38.RData\")\n",
    "\n",
    "make_ps_model <- function(cds_vcf, mut, AC_thresh = 1) {\n",
    "  singletons <- cds_vcf %>%\n",
    "    filter(AC == AC_thresh)\n",
    "  singletons_context <- table(singletons$mut)\n",
    "  context_n <- table(cds_vcf$mut)\n",
    "  prop_singletons <- data.frame('mutation' = names(singletons_context),\n",
    "                                'singletons' = c(singletons_context))\n",
    "  prop_n <- data.frame('mutation' = names(context_n),\n",
    "                       'n' = c(context_n))\n",
    "  prop_singletons <- left_join(prop_singletons, prop_n, by ='mutation')\n",
    "\n",
    "  prop_singletons <- left_join(prop_singletons, mut, by = 'mutation')\n",
    "  ps_lm <- lm(singletons/n ~ mu_snp, data = prop_singletons, weights = n)\n",
    "  return(ps_lm)\n",
    "\n",
    "}\n",
    "\n",
    "make_ps2_model <- function(cds_vcf, mu, AC_thresh = 1) {\n",
    "  variants <- split(cds_vcf, cds_vcf$group)\n",
    "  ps_models <- lapply(variants, function(x) make_ps_model(x, mu, AC_thresh))\n",
    "  return(ps_models)\n",
    "}\n",
    "\n",
    "annotate_methylation <- function(variants, mutation_table, methylation) {\n",
    "  variants <- left_join(variants, methylation)\n",
    "  # set missing methylation values to zero\n",
    "  variants$methyl_bins[is.na(variants$methyl_bins)] <- 0\n",
    "  fixed_context <- ifelse(variants$context %in% mutation_table$context,\n",
    "                          variants$context,\n",
    "                          chartr('ATGC','TACG',variants$context))\n",
    "  fixed_mut <- ifelse(variants$context %in% mutation_table$context,\n",
    "                      variants$mut_context,\n",
    "                      chartr('ATGC','TACG',variants$mut_context))\n",
    "\n",
    "  # amend mutation context for MAPS score calculation\n",
    "  variants$mut <- paste0(fixed_context,':',\n",
    "                         variants$methyl_bins,\n",
    "                         '>',\n",
    "                         fixed_mut)\n",
    "\n",
    "  variants$mut <- ifelse(variants$mut %in% mutation_table$mutation,\n",
    "                         variants$mut,\n",
    "                         gsub('[1|2]',\n",
    "                              '0',\n",
    "                              variants$mut))\n",
    "\n",
    "  variants$context <- fixed_context\n",
    "  variants$mut_context <- fixed_mut\n",
    "\n",
    "  # annotate variants for methylation adjustment\n",
    "  var_adjust <- mutation_table %>%\n",
    "    filter(methylation_level != 0) %>%\n",
    "    select(context) %>% unlist()\n",
    "\n",
    "  variants$group <- ifelse((variants$context %in% var_adjust),\n",
    "                           ifelse((variants$REF == 'C' & variants$ALT == 'T') | (variants$REF == 'G' & variants$ALT == 'A'),\n",
    "                                  'methyl_adjust',\n",
    "                                  'no_adjust'),\n",
    "                           'no_adjust')\n",
    "\n",
    "  return(variants)\n",
    "\n",
    "}\n",
    "bed_dir <- \"/disk2/G4_MS/frequency/MAPS/data\"\n",
    "setwd(bed_dir)\n",
    "# load variants from gnomAD v3.1\n",
    "vcf_syn <- data.table::fread(\"gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv\")\n",
    "colnames(vcf_syn) <- c('CHROM','POS','REF','ALT','AN','AC','AF','variant_type','lcr','vep')\n",
    "seq <- data.table::fread('gnomad.genomes.v3.1.sites.pass_snps.synonymous.seq',\n",
    "                        header = FALSE)\n",
    "seq <- toupper(seq$V2)\n",
    "vcf_syn$heptamer <- seq\n",
    "# filter variants\n",
    "vcf_syn <- vcf_syn %>%\n",
    "    filter(lcr != 1, AN >= 121849, CHROM != 'chrX', CHROM != 'chrY',variant_type == 'snv')\n",
    "vcf_syn$context <- substr(vcf_syn$heptamer, start = 3, stop = 5)\n",
    "vcf_syn$mut_context <- paste0(substr(vcf_syn$context, 1, 1),\n",
    "                      vcf_syn$ALT,\n",
    "                      substr(vcf_syn$context, 3, 3))\n",
    "\n",
    "# load mutation rate table\n",
    "mut <- data.table::fread('mutation_rate_methylation_bins.txt')\n",
    "mut <- mut %>%\n",
    "  mutate(ref_context = paste0(context, ':', methylation_level),\n",
    "         mut_context = paste0(substr(context, 1, 1),\n",
    "                              alt,\n",
    "                              substr(context, 3, 3)),\n",
    "         mutation = paste0(ref_context, '>', mut_context))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# annotate methylation levels\n",
    "vcf_syn <- annotate_methylation(vcf_syn, mut, combined_methylation_estimates)\n",
    "vcf_syn$methyl_bins <- as.numeric(as.character(vcf_syn$methyl_bins))\n",
    "# make MAPS model\n",
    "ps_lm_genomes <- make_ps2_model(vcf_syn, mut, AC_thresh = 1)\n",
    "lapply(ps_lm_genomes, function(x) summary(x))\n",
    "if (!dir.exists('RData')) {\n",
    "  dir.create('RData')\n",
    "}\n",
    "save(ps_lm_genomes, file = 'RData/ps_lm_genomes.RData')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76402a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /disk2/G4_MS/frequency/MAPS/data\n",
    "python maps_noG4_g4_data_get.py\n",
    "bedtools intersect -a /disk2/G4_MS/reference/gnomad.genomes.v3.1.sites.pass_snps.tsv.gz -b /disk2/G4_MS/reference/gencode38.cds_region.bed -wa -u|gzip > cds.gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\n",
    "python maps_synonynous.py\n",
    "python snp_trans.py\n",
    "bedtools sort -i gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv > gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv1\n",
    "mv gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv1 gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv\n",
    "awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv|bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab >gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.seq\n",
    "\n",
    "bedtools sort -i gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv > gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv1\n",
    "mv gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv1 gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv\n",
    "awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' gnomad.genomes.v3.1.sites.pass_snps.synonymous.tsv |bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab > gnomad.genomes.v3.1.sites.pass_snps.synonymous.seq\n",
    "\n",
    "tail -n +2 /disk2/G4_MS/promoter_select/promoter_info/promoter.annotion.tsv|bedtools sort -i - |bedtools merge -i -|bedtools intersect -a /disk2/G4_MS/reference/gnomad.genomes.v3.1.sites.pass_snps.tsv.gz -b -|bedtools intersect -a - -b /disk2/G4_MS/reference/gencode38.cds_region.bed -v |bedtools sort -i -|gzip > promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\n",
    "tail -n +2 /disk2/G4_MS/promoter_select/promoter_info/promoter_info.annotion.H3K4me3.tsv|bedtools sort -i - |bedtools merge -i -|bedtools intersect -a /disk2/G4_MS/reference/gnomad.genomes.v3.1.sites.pass_snps.tsv.gz -b -|bedtools intersect -a - -b /disk2/G4_MS/reference/gencode38.cds_region.bed -v |bedtools sort -i - |gzip > active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv.gz\n",
    "python snp_trans.py\n",
    "awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv|bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab > promoter_region_gnomad.genomes.v3.1.sites.pass_snps.seq\n",
    "awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv|bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab > active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.seq\n",
    "\n",
    "# bedtools sort -i gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv > gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv1\n",
    "# mv gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv1 gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv\n",
    "# awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv|bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab > gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv.seq\n",
    "\n",
    "bedtools sort -i gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv > gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv1\n",
    "mv gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv1 gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv\n",
    "awk -v OFS=\"\\t\" '{print$1,$2-4,$2+3}' gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv|bedtools getfasta -fi /usr/refence/hg38/hg38.fa -bed - -tab > gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv.seq\n",
    "\n",
    "Rscript maps.model.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68acb737",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Title     : TODO\n",
    "# Objective : TODO\n",
    "# Created by: guangyue\n",
    "# Created on: 2021/9/30\n",
    "# h3k4me3_stable_unstable.data.get.R\n",
    "\n",
    "library(GenomicRanges)\n",
    "library(dplyr)\n",
    "\n",
    "\n",
    "######################\n",
    "## Helper functions ##\n",
    "######################\n",
    "granges_from_vcf <- function(vcf) {\n",
    "\n",
    "  vcf$end <- vcf$POS\n",
    "  grange <- makeGRangesFromDataFrame(vcf, keep.extra.columns = TRUE, ignore.strand = TRUE,\n",
    "                                     seqnames.field = \"CHROM\", start.field = \"POS\", end.field = \"end\")\n",
    "  return(grange)\n",
    "\n",
    "}\n",
    "\n",
    "apply_qc <- function(vcf, coverage_thresh, SNV = TRUE, AF_thresh = Inf) {\n",
    "\n",
    "  \"\n",
    "  This function filters variants based on the specified QC metrics. If\n",
    "  SNV = TRUE, indels are removed. RF_PROB is the Random Forest Probability\n",
    "  that a called variant is a True Positive. The MacArthur Lab QC process\n",
    "  uses a RF probability of 0.4 Coverage threshold is the number of total\n",
    "  alleles that must have been observed (variant or non-variant) for a site.\n",
    "  In ExAC, the MacArthur Lab uses a cutoff of > 0.8 of the max AN for a site.\n",
    "  \"\n",
    "\n",
    "  # split X chromosome from others for AN threshold\n",
    "\n",
    "  thresh = 0.8*coverage_thresh\n",
    "\n",
    "  nonx <- vcf %>% filter(CHROM != \"chrX\" & CHROM != \"chrY\", AN > thresh)\n",
    "  combined_var <- nonx\n",
    "  combined_var$AF <- as.numeric(combined_var$AF)\n",
    "\n",
    "  if (SNV) {\n",
    "    combined_var <- combined_var %>% filter(nchar(REF) + nchar(ALT) == 2, lcr != 1, decoy != 1)\n",
    "  } else {\n",
    "    combined_var <- combined_var %>% filter(lcr != 1, decoy != 1)\n",
    "  }\n",
    "\n",
    "  combined_var <- combined_var %>% filter(AF < AF_thresh)\n",
    "  combined_var$AF <- as.numeric(combined_var$AF)\n",
    "\n",
    "  combined_var <- combined_var %>% filter(AC > 0)\n",
    "\n",
    "  return(combined_var)\n",
    "\n",
    "}\n",
    "\n",
    "bed_dir <- \"/disk2/G4_MS/frequency/MAPS/data\"\n",
    "reference_ranges <- list(g4_gruns = file.path(bed_dir, 'promoter_1000_eG4.gruns.bed'),\n",
    "                         g4_regions = file.path(bed_dir,'promoter_pG4.info.H3K4me3.tsv'),\n",
    "             stable_g4_gruns = file.path(bed_dir,'promoter_1000_stable_eG4.gruns.bed'),\n",
    "             unstable_g4_gruns = file.path(bed_dir,'promoter_1000_unstable_eG4.gruns.bed'),\n",
    "\n",
    "#                          g4tx_nog4_gruns = file.path(bed_dir,'protein_promoter_1000_noG4.pG4.gruns.g4_tx.bed'),\n",
    "                         nog4_gruns = file.path(bed_dir,'protein_promoter_1000_noG4.eG4.gruns.bed'),\n",
    "                         nog4_tx = file.path(bed_dir,'promoter_nopG4.info.H3K4me3.tsv')\n",
    "                         )\n",
    "\n",
    "reference_ranges$g4_gruns <- lapply(reference_ranges$g4_gruns, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"txid\", \"ensembl_gene_id\",\"oristrand\")))[[1]]\n",
    "reference_ranges$g4_regions <- lapply(reference_ranges$g4_regions, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"seq\", \"oristrand\",\"txid\",\"ensembl_gene_id\",\"oristrand_gene\")))[[1]]\n",
    "\n",
    "## reference_ranges$g4tx_nog4_gruns <- lapply(reference_ranges$g4tx_nog4_gruns, function (x) data.table::fread(\n",
    "#   x,col.names=c(\"chromosome\",'start','end',\"txid\", \"ensembl_gene_id\",\"oristrand\")))[[1]]\n",
    "reference_ranges$stable_g4_gruns <- lapply(reference_ranges$stable_g4_gruns, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"txid\", \"ensembl_gene_id\",\"oristrand\")))[[1]]\n",
    "reference_ranges$unstable_g4_gruns <- lapply(reference_ranges$unstable_g4_gruns, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"txid\",\"ensembl_gene_id\",\"oristrand\")))[[1]]\n",
    "\n",
    "reference_ranges$nog4_gruns <- lapply(reference_ranges$nog4_gruns, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"txid\", \"ensembl_gene_id\",\"oristrand\")))[[1]]\n",
    "reference_ranges$nog4_tx <- lapply(reference_ranges$nog4_tx, function (x) data.table::fread(\n",
    "  x,col.names=c(\"chromosome\",'start','end',\"prmtr_id\",\"ensembl_gene_id\", \"oristrand\",\"txid\")))[[1]]\n",
    "\n",
    "reference_ranges <- lapply(reference_ranges, function(x) x %>% filter(!chromosome == 'chrY'))\n",
    "reference_ranges <- lapply(reference_ranges, function(x) mutate(x,txid=substr(txid,1,15)))\n",
    "reference_ranges <- lapply(reference_ranges, function(x) mutate(x,ensembl_gene_id=substr(ensembl_gene_id,1,15)))\n",
    "reference_ranges <- lapply(reference_ranges, function(x) mutate(x,end=end-1))\n",
    "reference_granges <- lapply(reference_ranges, function(x) makeGRangesFromDataFrame(x,\n",
    "                                                                                   keep.extra.columns = TRUE,\n",
    "                                                                                   ignore.strand = TRUE,\n",
    "                                                                                   seqnames.field = \"chromosome\",\n",
    "                                                                                   start.field = \"start\",\n",
    "                                                                                   end.field = \"end\",\n",
    "                                                                                   starts.in.df.are.0based = FALSE))\n",
    "cds_ranges <- data.table::fread(file.path(bed_dir,'cds_region.used.bed'),\n",
    "                         col.names = c(\"chromosome\",'start','end',\"ensembl_gene_id\",\"gene_name\", \"oristrand\",\"txid\"))\n",
    "cds_ranges <- cds_ranges %>% filter(!chromosome == 'chrY')\n",
    "cds_ranges <- mutate(cds_ranges,ensembl_gene_id=substr(ensembl_gene_id,1,15))\n",
    "cds_ranges <- mutate(cds_ranges,txid=substr(txid,1,15))\n",
    "cds_ranges$end <- cds_ranges$end -1\n",
    "cds_ranges <- makeGRangesFromDataFrame(cds_ranges,\n",
    "                                       keep.extra.columns = TRUE,\n",
    "                                       ignore.strand = TRUE,\n",
    "                                       seqnames.field = 'chromosome',\n",
    "                                       start.field = 'start',\n",
    "                                       end.field = 'end',\n",
    "                                       starts.in.df.are.0based = FALSE)\n",
    "promoter_region <- data.table::fread(file.path(bed_dir,'promoter_info.annotion.H3K4me3.tsv'),\n",
    "                                     col.names = c(\"chromosome\",'start','end',\"prmtr_id\",\"ensembl_gene_id\", \"oristrand\",\"txid\")\n",
    ")\n",
    "promoter_region <- mutate(promoter_region,ensembl_gene_id=substr(ensembl_gene_id,1,15))\n",
    "promoter_region <- mutate(promoter_region,txid=substr(txid,1,15))\n",
    "promoter_region <- makeGRangesFromDataFrame(promoter_region,\n",
    "                                       keep.extra.columns = TRUE,\n",
    "                                       ignore.strand = TRUE,\n",
    "                                       seqnames.field = 'chromosome',\n",
    "                                       start.field = 'start',\n",
    "                                       end.field = 'end',\n",
    "                                       starts.in.df.are.0based = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c034df14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# promoter.MAPS.R\n",
    "library(dplyr)\n",
    "library(GenomicRanges)\n",
    "library(ggplot2)\n",
    "#library(ensembldb)\n",
    "\n",
    "setwd(\"/disk2/G4_MS/frequency/MAPS\")\n",
    "source(\"/disk2/G4_MS/frequency/MAPS/data/tss_stable_unstable.data.get.R\")\n",
    "load(\"data/RData/ps_lm_genomes.RData\")\n",
    "\n",
    "gtffile <- \"/disk2/G4_MS/frequency/MAPS/data/Homo_sapiens.GRCh38.104.gtf.gz\"\n",
    "DB.104 <- ensembldb::ensDbFromGtf(gtf = gtffile)\n",
    "EnsDb.Hsapiens.v104 <- ensembldb::EnsDb(DB.104)\n",
    "## cds rangeget\n",
    "mu <- data.table::fread(\"data/mutation_rate_methylation_bins.txt\")\n",
    "mu <- mu %>%\n",
    "  mutate(ref_context = paste0(context,\":\",methylation_level),\n",
    "         mut_context = paste0(substr(context,1,1),\n",
    "                              alt,\n",
    "                              substr(context,3,3)),\n",
    "         mutation = paste0(ref_context,\">\",mut_context))\n",
    "## Genes with alternativate pG4 isoform\n",
    "dlata_g4 <- unique(intersect(reference_ranges$nog4_tx$ensembl_gene_id,reference_ranges$g4_regions$ensembl_gene_id))\n",
    "constitutive_g4 <-unique(setdiff(reference_ranges$g4_regions$ensembl_gene_id,reference_ranges$nog4_tx$ensembl_gene_id))\n",
    "\n",
    "# load constraint table by gene\n",
    "gene_constraints <- data.table::fread(\"/disk2/G4_MS/reference/gnomad.v2.1.1.lof_metrics.by_gene.txt\")\n",
    "non_plof_genes <- gene_constraints[order(gene_constraints$oe_lof_upper,decreasing = TRUE),]\n",
    "non_plof_genes <- non_plof_genes[1:6568,]\n",
    "plof_genes <- gene_constraints[order(gene_constraints$oe_lof_upper,decreasing = FALSE),]\n",
    "plof_genes <- plof_genes[1:6568,]\n",
    "clinvar_genes <- data.table::fread(\"/disk2/G4_MS/reference/clinvar_20220115.vcf_annotated_af.tsv\")\n",
    "clinvar_genes <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,\n",
    "                                       keys=gsub(pattern = \":.*\",\"\",clinvar_genes$GENEINFO),\n",
    "                                       column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "cgcs <- data.table::fread(\"/disk2/G4_MS/reference//cancer_gene_census.tsv\")\n",
    "cgcs <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,keys=cgcs$GeneSymbol,column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "\n",
    "tsgs <- data.table::fread(\"/disk2/G4_MS/reference//tsgenes.tsv\")\n",
    "# tsgs <- tsgs[which(tsgs$GeneType=='protein-coding'),]\n",
    "tsgs <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,keys=tsgs$GeneSymbol,column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "\n",
    "# Load genome-wide methylation estimates\n",
    "load('data/methylation_hg38.RData')\n",
    "promoter_gnomad <- data.table::fread(\"data/promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv\")\n",
    "colnames(promoter_gnomad) <- c(\"CHROM\", \"POS\", \"REF\", \"ALT\", \"AN\", \"AC\", \"AF\")\n",
    "promoter_hept <- data.table::fread(\"data/promoter_region_gnomad.genomes.v3.1.sites.pass_snps.seq\", header = FALSE)\n",
    "promoter_tri <- substr(promoter_hept$V2,3,5)\n",
    "promoter_gnomad$context <- toupper(promoter_tri)\n",
    "promoter_gnomad$mut_context <- paste0(substr(promoter_gnomad$context,1,1),\n",
    "                                      promoter_gnomad$ALT,\n",
    "                                      substr(promoter_gnomad$context,3,3))\n",
    "promoter_gnomad$hept <- toupper(promoter_hept$V2)\n",
    "promoter_gnomad <- promoter_gnomad %>%  select(CHROM, POS, REF, ALT, AN, AC, AF, context, mut_context, hept) %>%\n",
    "  filter(CHROM != 'chrX', AN > 0.8*152312, AC >= 1)\n",
    "rm(promoter_hept)\n",
    "rm(promoter_tri)\n",
    "promoter_gnomad_grange <- granges_from_vcf(promoter_gnomad)\n",
    "### \n",
    "# combined_promoter_variants = promoter_gnomad_grange\n",
    "g1_allowed <- c('AGG',\n",
    "                'CCA',\n",
    "                'CCG',\n",
    "                'CCT',\n",
    "                'CGG',\n",
    "                'TGG')\n",
    "g2_allowed <- c('CCC','GGG')\n",
    "g3_allowed <- c('ACC',\n",
    "                'GCC',\n",
    "                'GGA',\n",
    "                'GGC',\n",
    "                'GGT',\n",
    "                'TCC')\n",
    "\n",
    "allowed <- list('pos1' = g1_allowed,\n",
    "                'pos2' = g2_allowed,\n",
    "                'pos3' = g3_allowed)\n",
    "\n",
    "######################\n",
    "## Helper Functions ##\n",
    "######################\n",
    "\n",
    "get_ps <- function(vcf_list, mu) {\n",
    "\n",
    "  n <- lapply(vcf_list, function(x) table(x$mutation))\n",
    "  ps <- lapply(vcf_list, function(x) table(x[x$AC == 1]$mutation))\n",
    "\n",
    "  ps_res <- mapply(function(x, y) cbind(x, y[match(names(x), names(y))]), x=n, y=ps, SIMPLIFY = FALSE)\n",
    "  ps_res <- lapply(ps_res, function(x) x[match(mu$mutation, rownames(x)),])\n",
    "\n",
    "  return(ps_res)\n",
    "}\n",
    "\n",
    "\n",
    "get_mu_avg <- function(ps_list, mu) {\n",
    "\n",
    "  res <- lapply(ps_list, function(x) sum((x*mu$mu_snp), na.rm = TRUE) / sum(x, na.rm = TRUE))\n",
    "\n",
    "\n",
    "  return(res)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "calc_maps <- function(raw_ps, ps_counts, predictions) {\n",
    "\n",
    "  obs <- unlist(raw_ps) * unlist(ps_counts)\n",
    "  exp <- unlist(predictions) * unlist(ps_counts)\n",
    "\n",
    "  maps <- (obs - exp) / unlist(ps_counts)\n",
    "\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "# MAPS 2 model:\n",
    "calc_maps2 <- function(variants, mu, ps_lm) {\n",
    "\n",
    "  variants <- data.frame(variants)\n",
    "\n",
    "  total_variants <- dim(variants)[[1]]\n",
    "\n",
    "  variants <- split(variants, variants$group)\n",
    "\n",
    "  if (length(variants) == 2) {\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else if (names(variants) == 'no_adjust') {\n",
    "\n",
    "    methyl_adjust <- c('exp_singletons.1' = 0,\n",
    "                       'obs_singletons' = 0)\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "  } else {\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "    no_adjust <- c('exp_singletons.1' = 0,\n",
    "                   'obs_singletons' = 0)\n",
    "  }\n",
    "  combined <- methyl_adjust + no_adjust\n",
    "\n",
    "  maps <- (combined['obs_singletons'] - combined['exp_singletons.1']) / total_variants\n",
    "  raw_p <- combined['obs_singletons'] / total_variants\n",
    "  se <-  sqrt(raw_p*(1-raw_p)/total_variants)\n",
    "  result <- c(\"MAPS\" = as.numeric(maps), \"SE\" = as.numeric(se), \"N\" = as.integer(total_variants))\n",
    "  return(result)\n",
    "\n",
    "}\n",
    "\n",
    "get_obs_exp_singletons <- function(variants, ps_lm) {\n",
    "\n",
    "  variants <- na.omit(variants)\n",
    "  mu_avg <- sum(variants$n * variants$mu_snp) / sum(variants$n)\n",
    "  exp_singletons <- sum(variants$n) * predict(ps_lm, data.frame('mu_snp' = mu_avg))\n",
    "  obs_singletons <- sum(variants$singletons)\n",
    "  return(c('exp_singletons' = exp_singletons,\n",
    "           'obs_singletons' = obs_singletons))\n",
    "\n",
    "}\n",
    "\n",
    "do_MAPS <- function(vcf_list, mu, ps_model) {\n",
    "  # determine proportion singletons\n",
    "  ps <- get_ps(vcf_list, mu)\n",
    "  # determine mu average for the variants + contexts\n",
    "  mu_avg <- get_mu_avg(ps, mu)\n",
    "  # extract raw proportion of singletons\n",
    "  raw_p <- lapply(vcf_list, function(x) length(x[x$AC==1])/length(x))\n",
    "  var_counts <- lapply(vcf_list, function(x) length(x))\n",
    "  se <- mapply(function(p, n ) sqrt(p*(1-p)/n), raw_p, var_counts)\n",
    "  # predicted ps:\n",
    "  ps_predict <- lapply(mu_avg, function(x) predict(ps_model, data.frame(mu = unlist(x))))\n",
    "  maps <- calc_maps(raw_p, var_counts, ps_predict)\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "\n",
    "annotate_grun_position <- function(grun_grange, promoter_region, width = 3) {\n",
    "\n",
    "  gruns <- grun_grange[width(grun_grange) == width]\n",
    "  gruns_df <- data.frame(gruns)\n",
    "  #gruns$strand <- gruns$oristrand\n",
    "  gruns_df$strand <- promoter_region$oristrand[match(gruns_df$txid, promoter_region$txid)]\n",
    "  # gruns_df$strand <- gruns_df$strand.1\n",
    "  # gruns_df <- gruns_df[,!colnames(gruns_df) %in% c(\"oristrand\",\"strand.1\")]\n",
    "  tile <- makeGRangesFromDataFrame(gruns_df)\n",
    "  tile <- tile(tile, width = 1) %>% as.data.frame()\n",
    "  pos_tiles <- tile %>% filter(strand == '+')\n",
    "  neg_tiles <- tile %>% filter(strand == '-')\n",
    "\n",
    "  pos_tiles$var_pos <- rep(x = c(1:width), dim(pos_tiles)[[1]]/width)\n",
    "  neg_tiles$var_pos <- rep(x = c(1:width), dim(neg_tiles)[[1]]/width)\n",
    "\n",
    "  tiles <- rbind(pos_tiles, neg_tiles)\n",
    "\n",
    "  pos1 <- tiles %>% filter(var_pos == 1)\n",
    "  pos2 <- tiles %>% filter(var_pos == 2)\n",
    "  pos3 <- tiles %>% filter(var_pos == 3)\n",
    "\n",
    "  res <- list(\"pos1\" = pos1,\n",
    "              \"pos2\" = pos2,\n",
    "              \"pos3\" = pos3)\n",
    "  res <- lapply(res, function(x) makeGRangesFromDataFrame(x, keep.extra.columns = TRUE))\n",
    "  return(res)\n",
    "\n",
    "}\n",
    "match_var_to_strand <- function(var_grange_first, var_grange_last, tx_grange) {\n",
    "\n",
    "  # this function annotates a variant as being either first or last in a trinucleotide G/C run\n",
    "  # based on the strand of the UTR transcript\n",
    "\n",
    "  pos_strand <- tx_grange[tx_grange$oristrand == '+']\n",
    "  neg_strand <- tx_grange[tx_grange$oristrand == '-']\n",
    "  pos_strand_first <- subsetByOverlaps(var_grange_first, pos_strand)\n",
    "  pos_strand_last <- subsetByOverlaps(var_grange_last, pos_strand)\n",
    "\n",
    "  neg_strand_first <- subsetByOverlaps(var_grange_last, neg_strand)\n",
    "  neg_strand_last <- subsetByOverlaps(var_grange_first, neg_strand)\n",
    "\n",
    "  first_vars <- c(pos_strand_first, neg_strand_first)\n",
    "  last_vars <- c(pos_strand_last, neg_strand_last)\n",
    "\n",
    "  return(list('pos1'=first_vars,\n",
    "              'pos3'=last_vars))\n",
    "\n",
    "}\n",
    "\n",
    "calc_maps2_bootstrap <- function(variants, mu, ps_lm) {\n",
    "\n",
    "  total_variants <- dim(variants)[[1]]\n",
    "\n",
    "  variants <- variants[sample(1:total_variants, total_variants, replace = TRUE),]\n",
    "\n",
    "  variants <- split(variants, variants$group)\n",
    "\n",
    "  if (length(variants) == 2) {\n",
    "\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else if (names(variants) == 'no_adjust') {\n",
    "\n",
    "    methyl_adjust <- c('exp_singletons.1' = 0,\n",
    "                       'obs_singletons' = 0)\n",
    "\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else {\n",
    "\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "\n",
    "    no_adjust <- c('exp_singletons.1' = 0,\n",
    "                   'obs_singletons' = 0)\n",
    "  }\n",
    "combined <- methyl_adjust + no_adjust\n",
    "\n",
    "  maps <- (combined['obs_singletons'] - combined['exp_singletons.1']) / total_variants\n",
    "\n",
    "  return(as.numeric(maps))\n",
    "\n",
    "}\n",
    "                \n",
    "bootstrap_maps2 <- function(variants, mu, ps_lm, ntimes = 10000) {\n",
    "\n",
    "  variants <- data.frame(variants)\n",
    "\n",
    "  variants <- variants %>%\n",
    "    select(AC, AN, group, mut)\n",
    "\n",
    "  res <- vector(\"list\", length = ntimes)\n",
    "\n",
    "  res <- lapply(res, function(x) calc_maps2_bootstrap(variants, mu, ps_lm))\n",
    "\n",
    "  return(unlist(res))\n",
    "\n",
    "}\n",
    "\n",
    "resample_MAPS <- function(vcf, mu, ps_model) {\n",
    "\n",
    "  n <- table(vcf$mut_context)\n",
    "  ps <- table(vcf[vcf$AC == 1]$mut_context)\n",
    "\n",
    "  ps <- cbind(n, ps[match(names(n), names(ps))])\n",
    "  ps <- ps[match(mu$mut_context, rownames(ps)),]\n",
    "\n",
    "  mu_average <- sum(ps*mu$mu_snp, na.rm =TRUE) / sum(ps, na.rm = TRUE)\n",
    "\n",
    "  raw_p <- length(vcf[vcf$AC==1])/length(vcf)\n",
    "  var_counts <- length(vcf)\n",
    "  ps_predict <- predict(ps_model, data.frame(mu = unlist(mu_average)))\n",
    "\n",
    "  obs <- unlist(raw_p) * unlist(var_counts)\n",
    "  exp <- unlist(ps_predict) * unlist(var_counts)\n",
    "  maps <- (obs-exp)/unlist(var_counts)\n",
    "\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "\n",
    "generate_bootstrap <- function(vcf, n, mu, ps_model) {\n",
    "\n",
    "  x <- length(vcf)\n",
    "\n",
    "  res <- rep(x, n)\n",
    "\n",
    "  samples <- lapply(res, function(x) sample(x, size = x, replace = TRUE))\n",
    "\n",
    "  maps_bootstrap <- lapply(samples, function(x) resample_MAPS(vcf[x], mu = mu, ps_model = ps_model))\n",
    "\n",
    "  return(unlist(maps_bootstrap))\n",
    "\n",
    "}\n",
    "                \n",
    "chunked_bootstrap <- function(vcf, n, mu, ps_model) {\n",
    "\n",
    "  chunksize <- as.integer(n/10)\n",
    "  chunk1 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk2 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk3 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk4 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk5 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk6 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk7 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk8 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk9 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk10 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                             mu = mu, ps_lm = ps_model)\n",
    "  result <- c(chunk1, chunk2, chunk3, chunk4, chunk5,\n",
    "              chunk6, chunk7, chunk8, chunk9, chunk10)\n",
    "  return(result)\n",
    "\n",
    "}\n",
    "annotate_methylation <- function(variants, mutation_table, methylation) {\n",
    "\n",
    "  variants <- left_join(variants, methylation)\n",
    "  # set missing methylation values to zero\n",
    "  variants$methyl_bins[is.na(variants$methyl_bins)] <- 0\n",
    "  fixed_context <- ifelse(variants$context %in% mutation_table$context,\n",
    "                          variants$context,\n",
    "                          chartr('ATGC','TACG',variants$context))\n",
    "  fixed_mut <- ifelse(variants$context %in% mutation_table$context,\n",
    "                      variants$mut_context,\n",
    "                      chartr('ATGC','TACG',variants$mut_context))\n",
    "\n",
    "  # amend mutation context for MAPS score calculation\n",
    "  variants$mut <- paste0(fixed_context,':',\n",
    "                         variants$methyl_bins,\n",
    "                         '>',\n",
    "                         fixed_mut)\n",
    "  variants$mut <- ifelse(variants$mut %in% mutation_table$mutation,\n",
    "                         variants$mut,\n",
    "                         gsub('[1|2]',\n",
    "                              '0',\n",
    "                              variants$mut))\n",
    "  variants$context <- fixed_context\n",
    "  variants$mut_context <- fixed_mut\n",
    "\n",
    "  # annotate variants for methylation adjustment\n",
    "  var_adjust <- mutation_table %>%\n",
    "    filter(methylation_level != 0) %>%\n",
    "    select(context) %>% unlist()\n",
    "  variants$group <- ifelse((variants$context %in% var_adjust),\n",
    "                           ifelse((variants$REF == 'C' & variants$ALT == 'T') | (variants$REF == 'G' & variants$ALT == 'A'),\n",
    "                                  'methyl_adjust',\n",
    "                                  'no_adjust'),\n",
    "                           'no_adjust')\n",
    "\n",
    "  return(variants)\n",
    "\n",
    "}\n",
    "\n",
    "##############\n",
    "## Analysis ##\n",
    "##############\n",
    "# annotate methylation context for MAPS analysis\n",
    "# correct gruns\n",
    "reference_granges$g4_gruns <- subsetByOverlaps(reference_granges$g4_gruns, promoter_region)\n",
    "## combined_promoter_variants = promoter_gnomad_grange\n",
    "promoter_gnomad_grange_df <- data.frame(promoter_gnomad_grange) %>% mutate(CHROM = seqnames, POS = start)\n",
    "promoter_gnomad_grange_df <- annotate_methylation(data.frame(promoter_gnomad_grange_df),\n",
    "                                                  mu,\n",
    "                                                  combined_methylation_estimates)\n",
    "promoter_gnomad_grange <- promoter_gnomad_grange_df %>% select(-c(CHROM, POS)) %>%\n",
    "  makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "\n",
    "# subset pG4s by alternative vs. constitutive\n",
    "promoter_delta_pg4 <- reference_granges$g4_gruns[reference_granges$g4_gruns$ensembl_gene_id %in% dlata_g4]\n",
    "promoter_const_pg4 <- reference_granges$g4_gruns[!reference_granges$g4_gruns$ensembl_gene_id %in% dlata_g4]\n",
    "                           \n",
    "promoter_stable_pg4 <- reference_granges$stable_g4_gruns\n",
    "#promoter_stable_pg4 <- promoter_stable_pg4[!promoter_stable_pg4$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "promoter_unstable_pg4 <-reference_granges$unstable_g4_gruns\n",
    "# promoter_unstable_pg4 <- promoter_unstable_pg4[!promoter_unstable_pg4$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "# Annotate grun positions (G1, G2, G3) and remove sequence contexts that are mismatched\n",
    "promoter_g4_gruns_positional <- annotate_grun_position(reference_granges$g4_gruns,promoter_region)\n",
    "promoter_g4_vars <- lapply(promoter_g4_gruns_positional, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for delta pG4\n",
    "promoter_delta_pg4 <- annotate_grun_position(promoter_delta_pg4,promoter_region)\n",
    "promoter_delta_pg4 <- lapply(promoter_delta_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for constitutive pG4\n",
    "promoter_const_pg4 <- annotate_grun_position(promoter_const_pg4,promoter_region)\n",
    "promoter_const_pg4 <- lapply(promoter_const_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))                     \n",
    "\n",
    "# repeat for stable pG4\n",
    "promoter_stable_pg4 <- annotate_grun_position(promoter_stable_pg4,promoter_region)\n",
    "promoter_stable_pg4 <- lapply(promoter_stable_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for unstable pG4\n",
    "promoter_unstable_pg4 <- annotate_grun_position(promoter_unstable_pg4,promoter_region)\n",
    "promoter_unstable_pg4 <- lapply(promoter_unstable_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# combine variant sets and perform MAPS\n",
    "promoter_combined <- c('G1' = promoter_g4_vars$pos1,\n",
    "                  'G2' = promoter_g4_vars$pos2,\n",
    "                  'G3' = promoter_g4_vars$pos3\n",
    ")\n",
    "promoter_combined_delta <- c('G1' = promoter_delta_pg4$pos1,\n",
    "                             'G2' = promoter_delta_pg4$pos2,\n",
    "                             'G3' = promoter_delta_pg4$pos3\n",
    ")\n",
    "promoter_combined_const <- c('G1' = promoter_const_pg4$pos1,\n",
    "                             'G2' = promoter_const_pg4$pos2,\n",
    "                             'G3' = promoter_const_pg4$pos3\n",
    ")\n",
    "###\n",
    "promoter_combined_stable <- c('G1' = promoter_stable_pg4$pos1,\n",
    "                             'G2' = promoter_stable_pg4$pos2,\n",
    "                             'G3' = promoter_stable_pg4$pos3\n",
    ")\n",
    "promoter_combined_unstable <- c('G1' = promoter_unstable_pg4$pos1,\n",
    "                             'G2' = promoter_unstable_pg4$pos2,\n",
    "                             'G3' = promoter_unstable_pg4$pos3\n",
    ")\n",
    "# Evaluate MAPS\n",
    "maps_pg4_combined <- lapply(promoter_combined, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_delta <- lapply(promoter_combined_delta, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_const<- lapply(promoter_combined_const, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_stable <- lapply(promoter_combined_stable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_unstable <- lapply(promoter_combined_unstable, function(x) calc_maps2(x, mu, ps_lm_genomes))                                       \n",
    "\n",
    "######################################################################\n",
    "## repeat for GGG/CCC repeats to control for local sequence context ##\n",
    "######################################################################\n",
    "\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_gnomad_grange, promoter_combined$G1, invert = TRUE)\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, promoter_combined$G2, invert = TRUE)\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, promoter_combined$G3, invert = TRUE)\n",
    "# remove CDS\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, cds_ranges, invert = TRUE)\n",
    "# remove CpG sites\n",
    "promoter_combined_npg4_central <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 2, 6))\n",
    "promoter_combined_npg4_central <- promoter_combined_npg4[promoter_combined_npg4_central]\n",
    "\n",
    "# Map 1st and 3rd positions for non-pG4 sequences:\n",
    "promoter_region_df <- data.frame(promoter_region)\n",
    "promoter_region_df$strand <- promoter_region_df$oristrand\n",
    "# promoter_region_df <- promoter_region_df[,!colnames(promoter_region_df) %in% c(\"oristrand\")]\n",
    "promoter_region_df <- makeGRangesFromDataFrame(promoter_region_df,\n",
    "                                               keep.extra.columns = TRUE,\n",
    "                                               seqnames.field = 'seqnames',\n",
    "                                               start.field = 'start',\n",
    "                                               end.field = 'end')\n",
    "\n",
    "promoter_combined_npg4_first <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 3, 7))\n",
    "promoter_combined_npg4_first <- promoter_combined_npg4[promoter_combined_npg4_first]\n",
    "\n",
    "promoter_combined_npg4_last <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 1, 5))\n",
    "promoter_combined_npg4_last <- promoter_combined_npg4[promoter_combined_npg4_last]\n",
    "\n",
    "first_last_npg4 <- match_var_to_strand(promoter_combined_npg4_first, promoter_combined_npg4_last,\n",
    "                                       promoter_region_df)\n",
    "\n",
    "promoter_nog4_combined <- list('pos1' = first_last_npg4$pos1,\n",
    "                      'pos2' = promoter_combined_npg4_central,\n",
    "                      'pos3' = first_last_npg4$pos3)\n",
    "# All non-pG4 GGG / CCC contexts\n",
    "maps_nong4_first_last <- lapply(promoter_nog4_combined, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_all_promoter <- list('all'=calc_maps2(promoter_gnomad_grange, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 delta GGG /CCC contexts\n",
    "promoter_nog4_combined_delta <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_delta <- lapply(promoter_nog4_combined_delta, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_delta_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    ")\n",
    "all_promoter_delta_variants <- subsetByOverlaps(all_promoter_delta_variants,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    "                                                )\n",
    "maps_all_promoter_delta <- list('all'=calc_maps2(all_promoter_delta_variants, mu, ps_lm_genomes))\n",
    "                                \n",
    "# non-pG4 const GGG /CCC contexts\n",
    "promoter_nog4_combined_const <- lapply(promoter_nog4_combined, function(x)\n",
    "  subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    "))\n",
    "promoter_nog4_combined_const <- lapply(promoter_nog4_combined_const, function(x)\n",
    "  subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    "  )\n",
    ")\n",
    "\n",
    "maps_promoter_nog4_const <- lapply(promoter_nog4_combined_const, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "# select variants that are not in delta pG4 genes' promoter\n",
    "const_all_promoter_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% dlata_g4])\n",
    "# subset these variants by those that are in pG4 genes to isolate \"constitutive\" pG4s\n",
    "const_all_promoter_variants <- subsetByOverlaps(const_all_promoter_variants,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    ")\n",
    "maps_all_promoter_const <- list('all'=calc_maps2(const_all_promoter_variants, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 stable GGG/CCC contexts\n",
    "promoter_nog4_combined_stable <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "promoter_nog4_combined_stable <- lapply(promoter_nog4_combined_stable, function(x) subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_stable <- lapply(promoter_nog4_combined_stable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_stable_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           reference_granges$stable_g4_gruns\n",
    ")\n",
    "#all_promoter_stable_variants <- subsetByOverlaps(all_promoter_stable_variants,\n",
    "#                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "#)\n",
    "maps_all_promoter_stable <- list('all'=calc_maps2(all_promoter_stable_variants, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 unstable GGG/CCC contexts\n",
    "promoter_nog4_combined_unstable <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "promoter_nog4_combined_unstable <- lapply(promoter_nog4_combined_unstable, function(x) subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_unstable <- lapply(promoter_nog4_combined_unstable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_unstable_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           reference_granges$unstable_g4_gruns)\n",
    "#all_promoter_unstable_variants <- subsetByOverlaps(all_promoter_unstable_variants,\n",
    "#                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "#)\n",
    "maps_all_promoter_unstable <- list('all'=calc_maps2(all_promoter_unstable_variants, mu, ps_lm_genomes)) \n",
    "                                      \n",
    "####################################################################\n",
    "## Evaluate MAPS for variants in top 1/3 of most constrained genes ##\n",
    "#####################################################################\n",
    "## promoter_g4_gruns = combined_g4_gruns\n",
    "promoter_g4_gruns <- reference_granges$g4_gruns[width(reference_granges$g4_gruns) == 3]\n",
    "promoter_g4_gruns$plof <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% plof_genes$gene_id,\n",
    "                                 TRUE,\n",
    "                                 FALSE)\n",
    "\n",
    "promoter_combined_g4_plof_gruns <- lapply(promoter_combined,\n",
    "                                 function(x) subsetByOverlaps(\n",
    "                                   x, promoter_g4_gruns[promoter_g4_gruns$plof == TRUE])\n",
    "                                 )\n",
    "\n",
    "# overlap gnomAD variants\n",
    "combined_g4_plof_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$plof == TRUE]))\n",
    "\n",
    "#combined_g4_nonplof_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$non_plof == TRUE]))\n",
    "\n",
    "# evaluate MAPS for top third most constrained UTR\n",
    "maps_plof_gruns <-  lapply(combined_g4_plof_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "# Evaluate all variants in top 1/3 most constrained UTR\n",
    "promoter_reference_grange <- data.frame(promoter_region) %>%\n",
    "  select(txid,\n",
    "         ensembl_gene_id,\n",
    "         start,\n",
    "         end,\n",
    "         oristrand,\n",
    "         seqnames)\n",
    "promoter_reference_grange$strand <- promoter_reference_grange$oristrand\n",
    "promoter_reference_grange$top_third <- ifelse(promoter_reference_grange$ensembl_gene_id %in% plof_genes$gene_id, TRUE, FALSE)\n",
    "promoter_reference_grange$bottom_third <- ifelse(promoter_reference_grange$ensembl_gene_id %in% non_plof_genes$gene_id, TRUE,FALSE)\n",
    "promoter_reference_grange$clinvar_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% clinvar_genes, TRUE, FALSE)\n",
    "promoter_reference_grange$cgcs_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% cgcs, TRUE, FALSE)\n",
    "promoter_reference_grange$tsgs_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% tsgs, TRUE, FALSE)\n",
    "promoter_reference_grange <- promoter_reference_grange %>% makeGRangesFromDataFrame(start.field = 'start',\n",
    "                                                                            end.field = 'end',\n",
    "                                                                            seqnames.field = 'seqnames', keep.extra.columns =TRUE)\n",
    "\n",
    "# define variant subsets\n",
    "all_promoter_plof <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$top_third == TRUE])\n",
    "all_promoter_nonplof <- subsetByOverlaps(promoter_gnomad_grange,promoter_reference_grange[promoter_reference_grange$bottom_third == TRUE])\n",
    "all_promoter_clinvar_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$clinvar_signal == TRUE])\n",
    "all_promoter_cgcs_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$cgcs_signal == TRUE])\n",
    "all_promoter_tsgs_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$tsgs_signal == TRUE])\n",
    "\n",
    "nog4_combined_plof <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$top_third == TRUE]))\n",
    "nog4_combined_clinvar_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$clinvar_signal == TRUE]))\n",
    "nog4_combined_cgcs_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$cgcs_signal == TRUE]))\n",
    "nog4_combined_tsgs_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$tsgs_signal == TRUE]))\n",
    "\n",
    "\n",
    "promoter_g4_gruns$clinvar_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% clinvar_genes, TRUE, FALSE)\n",
    "promoter_g4_gruns$cgcs_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% cgcs, TRUE, FALSE)\n",
    "promoter_g4_gruns$tsgs_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% tsgs, TRUE, FALSE)\n",
    "\n",
    "#####################################\n",
    "## evaluate MAPS for ClinVar genes ##\n",
    "#####################################\n",
    "maps_plof_all_variants <- list('all_promoter_plof' = calc_maps2(all_promoter_plof, mu, ps_lm_genomes))\n",
    "\n",
    "maps_nog4_combined_plof <- lapply(nog4_combined_plof, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "maps_clinvar_signal_all_variants <- list('clinvar' = calc_maps2(all_promoter_clinvar_signal, mu, ps_lm_genomes))\n",
    "maps_cgcs_signal_all_variants <- list('cgcs' = calc_maps2(all_promoter_cgcs_signal, mu, ps_lm_genomes))\n",
    "maps_tsgs_signal_all_variants <- list('tsgs' = calc_maps2(all_promoter_tsgs_signal, mu, ps_lm_genomes))\n",
    "\n",
    "combined_g4_clinvar_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$clinvar_signal == TRUE]))\n",
    "combined_g4_cgcs_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$cgcs_signal == TRUE]))\n",
    "combined_g4_tsgs_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$tsgs_signal == TRUE]))\n",
    "\n",
    "maps_clinvar_pg4 <- lapply(combined_g4_clinvar_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_pg4 <- lapply(combined_g4_cgcs_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_pg4 <- lapply(combined_g4_tsgs_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "maps_clinvar_nong4 <- lapply(nog4_combined_clinvar_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_nong4 <- lapply(nog4_combined_cgcs_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_nong4 <- lapply(nog4_combined_tsgs_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "\n",
    "combined_g4_clinvar_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_clinvar_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "combined_g4_cgcs_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_cgcs_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "combined_g4_tsgs_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_tsgs_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "\n",
    "\n",
    "maps_clinvar_pg4_delta <- lapply(combined_g4_clinvar_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_pg4_delta <- lapply(combined_g4_cgcs_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_pg4_delta <- lapply(combined_g4_tsgs_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "# Generate MAPS estimates for missense and putative loss of function-causing variants in gnomAD\n",
    "missense <- data.table::fread('data/gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv',header=FALSE)\n",
    "colnames(missense) <- c('CHROM','POS','REF','ALT','AN','AC','AF','type','other', 'vep')\n",
    "heptamers <- data.table::fread(\"data/gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.seq\", header = FALSE)\n",
    "missense$heptamers <- toupper(heptamers$V2)\n",
    "missense$context <- toupper(substr(heptamers$V2, 3, 5))\n",
    "\n",
    "rm(heptamers)\n",
    "                              \n",
    "an_adj <- 0.8*152312\n",
    "missense <- missense %>% filter(AC != 0,\n",
    "                                AN >= an_adj, CHROM != \"chrX\") %>%\n",
    "                        select(-c(type, other, vep))\n",
    "\n",
    "missense$mut_context <- paste0(substr(missense$context, 1, 1),\n",
    "                               missense$ALT,\n",
    "                               substr(missense$context, 3, 3))\n",
    "\n",
    "missense$CHROM <- as.character(missense$CHROM)\n",
    "missense <- annotate_methylation(missense, mu, combined_methylation_estimates)\n",
    "plof <- data.table::fread(\"data/gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv\",\n",
    "                          header=FALSE,\n",
    "                          col.names = c('CHROM','POS','REF','ALT','AN','AC','AF','type','other', 'vep')\n",
    ")\n",
    "heptamers <- data.table::fread('data/gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv.seq', header = FALSE)\n",
    "plof$heptamers <- heptamers$V2 %>% toupper()\n",
    "rm(heptamers)\n",
    "plof <- plof %>% filter(AC >= 1,\n",
    "                        AN >= 0.8*152312, CHROM != \"chrX\",\n",
    "                        nchar(REF) == 1, nchar(ALT) == 1) %>%\n",
    "  select(-c(type, other, vep))\n",
    "\n",
    "plof$context <- substr(plof$heptamers, 3,5)\n",
    "\n",
    "plof$mut_context <- paste0(substr(plof$context, 1, 1),\n",
    "                           plof$ALT,\n",
    "                           substr(plof$context, 3, 3))\n",
    "\n",
    "plof$CHROM <- as.character(plof$CHROM)\n",
    "plof <- annotate_methylation(plof, mu, combined_methylation_estimates)\n",
    "####\n",
    "cds_varsets <- list('missense_all' = missense,\n",
    "                    'lof_all' = plof)\n",
    "maps_cds_markers <- lapply(cds_varsets, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "##############################\n",
    "## Generate bootstrapped CI ##\n",
    "##############################\n",
    "# G4_combined bootstraps\n",
    "g4_combined_bootstraps <- lapply(promoter_combined, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "save(g4_combined_bootstraps, file = 'RData/g4_positional_maps_bootstraps_ncpg.RData')\n",
    "lapply(g4_combined_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/g4_positional_maps_bootstraps_ncpg.RData')\n",
    "\n",
    "G4 delta bootstraps\n",
    "g4_combined_delta_bootstraps <- lapply(promoter_combined_delta, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000)) #lapply(utr_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(g4_combined_delta_bootstraps, file = 'RData/g4_combined_delta_bootstraps.RData')\n",
    "lapply(g4_combined_delta_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/g4_combined_delta_bootstraps.RData\")\n",
    "\n",
    "# G4 const bootstraps\n",
    "g4_combined_const_bootstraps <- lapply(promoter_combined_const, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(utr_combined_const, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(g4_combined_const_bootstraps, file = 'RData/g4_combined_const_bootstraps.RData')\n",
    "lapply(g4_combined_const_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/g4_combined_const_bootstraps.RData\")\n",
    "       \n",
    "# G4 stable bootstraps\n",
    "g4_combined_stable_bootstraps <- lapply(promoter_combined_stable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000)) \n",
    "save(g4_combined_stable_bootstraps, file = 'RData/g4_combined_stable_bootstraps.RData')\n",
    "#lapply(g4_combined_stable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "#load(\"RData/g4_combined_stable_bootstraps.RData\")\n",
    "\n",
    "# G4 unstable bootstraps\n",
    "g4_combined_unstable_bootstraps <- lapply(promoter_combined_unstable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "save(g4_combined_unstable_bootstraps, file = 'RData/g4_combined_unstable_bootstraps.RData')\n",
    "#lapply(g4_combined_unstable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "#load(\"RData/g4_combined_unstable_bootstraps.RData\")\n",
    "\n",
    "# # non-G4 G bootstraps\n",
    "nog4_combined_bootstraps <- lapply(promoter_nog4_combined, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_bootstraps, file = 'RData/nog4_combined_bootstraps.RData')\n",
    "#lapply(nog4_combined_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "load(\"RData/nog4_combined_bootstraps.RData\")\n",
    "\n",
    "# # non-G4 delta-G bootstraps\n",
    "nog4_combined_delta_bootstraps <-  lapply(promoter_nog4_combined_delta, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_delta_bootstraps, file = 'RData/nog4_combined_delta_bootstraps.RData')\n",
    "lapply(nog4_combined_delta_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_delta_bootstraps.RData')\n",
    "\n",
    "# # non-G4 const-G bootstraps\n",
    "nog4_combined_const_bootstraps <-  lapply(promoter_nog4_combined_const, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_const, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_const_bootstraps, file = 'RData/nog4_combined_const_bootstraps.RData')\n",
    "#lapply(nog4_combined_const_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_const_bootstraps.RData')\n",
    "\n",
    "# non-G4 stable-G bootstraps\n",
    "nog4_combined_stable_bootstraps <-  lapply(promoter_nog4_combined_stable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_stable_bootstraps, file = 'RData/nog4_combined_stable_bootstraps.RData')\n",
    "#lapply(nog4_combined_stable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "#load('RData/nog4_combined_stable_bootstraps.RData')\n",
    "\n",
    "# non-G4 unstable-G bootstraps\n",
    "nog4_combined_unstable_bootstraps <-  lapply(promoter_nog4_combined_unstable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_unstable_bootstraps, file = 'RData/nog4_combined_unstable_bootstraps.RData')\n",
    "#lapply(nog4_combined_unstable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "#load('RData/nog4_combined_unstable_bootstraps.RData')       \n",
    "       \n",
    "# # g4 in PLOF genes bootstrap\n",
    "combined_g4_plof_gruns_bootstrap <-  lapply(promoter_combined_g4_plof_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_plof_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_plof_gruns_bootstrap, file = 'RData/combined_g4_plof_gruns_bootstraps.RData')\n",
    "#lapply(combined_g4_plof_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "load(\"RData/combined_g4_plof_gruns_bootstraps.RData\")\n",
    "\n",
    "# # nonG4 in PLOF genes bootstrap\n",
    "combined_npg4_plof_gruns_bootstrap <-  lapply(nog4_combined_plof, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_plof, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_npg4_plof_gruns_bootstrap, file = 'RData/combined_npg4_plof_gruns_bootstrap.RData')\n",
    "lapply(combined_npg4_plof_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_npg4_plof_gruns_bootstrap.RData')\n",
    "# #\n",
    "# G4 in disease gene bootstrap\n",
    "combined_g4_clinvar_signal_bootstrap <-  lapply(combined_g4_clinvar_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_clinvar_signal_bootstrap, file = 'RData/combined_g4_clinvar_signal_bootstrap.RData')\n",
    "lapply(combined_g4_clinvar_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_clinvar_signal_bootstrap.RData')\n",
    "\n",
    "# # G4 in cgcs gene bootstrap\n",
    "combined_g4_cgcs_signal_bootstrap <-  lapply(combined_g4_cgcs_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_cgcs_signal_bootstrap, file = 'RData/combined_g4_cgcs_signal_bootstrap.RData')\n",
    "lapply(combined_g4_cgcs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_cgcs_signal_bootstrap.RData')\n",
    "\n",
    "# # G4 in disease gene bootstrap\n",
    "combined_g4_tsgs_signal_bootstrap <-  lapply(combined_g4_tsgs_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_tsgs_signal_bootstrap, file = 'RData/combined_g4_tsgs_signal_bootstrap.RData')\n",
    "lapply(combined_g4_tsgs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_tsgs_signal_bootstrap.RData')\n",
    "\n",
    "# nonG4 in disease gene bootstrap\n",
    "nog4_combined_clinvar_signal_bootstrap <-  lapply(nog4_combined_clinvar_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm= ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_clinvar_signal_bootstrap, file = 'RData/nog4_combined_clivar_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_clinvar_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_clivar_signal_bootstrap.RData')\n",
    "# #\n",
    "# nonG4 in disease gene bootstrap\n",
    "nog4_combined_cgcs_signal_bootstrap <-  lapply(nog4_combined_cgcs_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_cgcs_signal_bootstrap, file = 'RData/nog4_combined_cgcs_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_cgcs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "load('RData/nog4_combined_cgcs_signal_bootstrap.RData')\n",
    "# #\n",
    "# nonG4 in disease gene bootstrap\n",
    "nog4_combined_tsgs_signal_bootstrap <-  lapply(nog4_combined_tsgs_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_tsgs_signal_bootstrap, file = 'RData/nog4_combined_tsgs_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_tsgs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_tsgs_signal_bootstrap.RData')\n",
    "\n",
    "# delta g4 in disease genes bootstrap\n",
    "combined_g4_clinvar_signal_delta_gruns_bootstrap <-  lapply(combined_g4_clinvar_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_clinvar_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_clinvar_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_clinvar_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_clinvar_signal_delta_gruns_bootstrap.RData')\n",
    "\n",
    "# delta g4 in cgcs genes bootstrap\n",
    "combined_g4_cgcs_signal_delta_gruns_bootstrap <-  lapply(combined_g4_cgcs_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_cgcs_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_cgcs_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_cgcs_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_cgcs_signal_delta_gruns_bootstrap.RData')\n",
    "# \n",
    "# # delta g4 in tsgs genes bootstrap\n",
    "combined_g4_tsgs_signal_delta_gruns_bootstrap <-  lapply(combined_g4_tsgs_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_tsgs_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_tsgs_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_tsgs_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_tsgs_signal_delta_gruns_bootstrap.RData')\n",
    "\n",
    "# ALL promoter variants bootstrap\n",
    "promoter_gnomad_grange_bootstrap <- chunked_bootstrap(promoter_gnomad_grange, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(promoter_gnomad_grange_bootstrap, file = 'RData/promoter_gnomad_grange_bootstrap.RData')\n",
    "quantile(promoter_gnomad_grange_bootstrap, c(0.05, 0.95))\n",
    "load('RData/promoter_gnomad_grange_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants delta pG4 bootstrap\n",
    "combined_promoter_delta_variants_bootstrap <- chunked_bootstrap(all_promoter_delta_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_delta_variants_bootstrap, file = 'RData/combined_promoter_delta_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_delta_variants_bootstrap, c(0.05, 0.95))\n",
    "load('RData/combined_promoter_delta_variants_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants const pG4 bootstrap\n",
    "combined_promoter_const_variants_bootstrap <- chunked_bootstrap(const_all_promoter_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_const_variants_bootstrap, file = 'RData/combined_promoter_const_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_const_variants_bootstrap, c(0.05, 0.95))\n",
    "load(\"RData/combined_promoter_const_variants_bootstrap.RData\")\n",
    "\n",
    "# ALL promoter variants stable pG4 bootstrap\n",
    "combined_promoter_stable_variants_bootstrap <- chunked_bootstrap(all_promoter_stable_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_stable_variants_bootstrap, file = 'RData/combined_promoter_stable_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_stable_variants_bootstrap, c(0.05, 0.95))\n",
    "load('RData/combined_promoter_stable_variants_bootstrap.RData')\n",
    "\n",
    "# ALL promoter variants unstable pG4 bootstrap\n",
    "combined_promoter_unstable_variants_bootstrap <- chunked_bootstrap(all_promoter_unstable_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_unstable_variants_bootstrap, file = 'RData/combined_promoter_unstable_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_unstable_variants_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/combined_promoter_unstable_variants_bootstrap.RData')\n",
    "\n",
    "# # All promoter variants plof bootstrap\n",
    "all_promoter_plof_bootstrap <- chunked_bootstrap(all_promoter_plof, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_plof_bootstrap, file = 'RData/all_promoter_plof_bootstrap.RData')\n",
    "quantile(all_promoter_plof_bootstrap, c(0.05, 0.95))\n",
    "load('RData/all_promoter_plof_bootstrap.RData')\n",
    "\n",
    " # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_clinvar_dx_bootstrap <- chunked_bootstrap(all_promoter_clinvar_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_clinvar_dx_bootstrap, file = 'RData/all_promoter_clinvar_dx_bootstrap.RData')\n",
    "quantile(all_promoter_clinvar_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_clinvar_dx_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_cgcs_dx_bootstrap <- chunked_bootstrap(all_promoter_cgcs_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_cgcs_dx_bootstrap, file = 'RData/all_promoter_cgcs_dx_bootstrap.RData')\n",
    "quantile(all_promoter_cgcs_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_cgcs_dx_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_tsgs_dx_bootstrap <- chunked_bootstrap(all_promoter_tsgs_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_tsgs_dx_bootstrap, file = 'RData/all_promoter_tsgs_dx_bootstrap.RData')\n",
    "quantile(all_promoter_tsgs_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_tsgs_dx_bootstrap.RData')\n",
    "                                      \n",
    "                     \n",
    "confidence_intervals <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                             'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                             'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                             'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                             'Top-third SNV' = all_promoter_plof_bootstrap,\n",
    "                             'Top-third pG4 G1' = combined_g4_plof_gruns_bootstrap$G1,\n",
    "                             'Top-third pG4 G2' = combined_g4_plof_gruns_bootstrap$G2,\n",
    "                             'Top-third pG4 G3' = combined_g4_plof_gruns_bootstrap$G3,\n",
    "                             'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                             'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                             'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                             'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                             'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                             'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                             'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                             'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "\n",
    "                             'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                             'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                             'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                             'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3\n",
    "                             )\n",
    "\n",
    "confidence_intervals <- sapply(confidence_intervals, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                               maps_pg4_combined$G1['MAPS'],\n",
    "                               maps_pg4_combined$G2['MAPS'],\n",
    "                               maps_pg4_combined$G3['MAPS'],\n",
    "                               maps_plof_all_variants$all_promoter_plof['MAPS'],\n",
    "                               maps_plof_gruns$G1['MAPS'],\n",
    "                               maps_plof_gruns$G2['MAPS'],\n",
    "                               maps_plof_gruns$G3['MAPS'],\n",
    "                               maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                               maps_clinvar_pg4$G1['MAPS'],\n",
    "                               maps_clinvar_pg4$G2['MAPS'],\n",
    "                               maps_clinvar_pg4$G3['MAPS'],\n",
    "                               maps_cgcs_signal_all_variants$cgcs[\"MAPS\"],\n",
    "                               maps_cgcs_pg4$G1['MAPS'],\n",
    "                               maps_cgcs_pg4$G2[\"MAPS\"],\n",
    "                               maps_cgcs_pg4$G3[\"MAPS\"],\n",
    "                               maps_tsgs_signal_all_variants$tsgs[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G1[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G2[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G3[\"MAPS\"]\n",
    "                               )\n",
    "\n",
    "plot_df <- confidence_intervals\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"LOEUF Top 1/3\", \"LOEUF Top 1/3\", \"LOEUF Top 1/3\", \"LOEUF Top 1/3\",\n",
    "                             \"ClinVar Genes\", \"ClinVar Genes\", \"ClinVar Genes\", \"ClinVar Genes\",\n",
    "                             \"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\n",
    "                             \"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_plof),\n",
    "               sapply(combined_g4_plof_gruns, function(x) length(x)),\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_gruns, function(x) length(x))\n",
    ")\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "\n",
    "# tiff('figures/MAPS.tiff', units = 'in', width = 9, height = 6, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "#  geom_segment(x = 10, xend = 10, y = plot_df['ClinVar pG4 G1',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "#  geom_segment(x = 8, xend = 8, y = plot_df['Top-third pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, 0.10)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'red') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.4, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.0, y = 0.096, label = \"pLOF\", color = 'red', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'LOEUF Top 1/3', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 3)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 3)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 3)\n",
    "ggsave('figures/MAPS.tiff', units = 'in', width = 9, height = 6, device=tiff())\n",
    "ggsave('figures/MAPS.pdf', width = 9, height = 6,device=pdf())\n",
    "dev.off()\n",
    "\n",
    "###################################\n",
    "## Plot delta pG4 vs. const. pG4 ##\n",
    "###################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                               'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                               'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                               'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                               'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                               'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                               'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                               'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "                               'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                               'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                               'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                               'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3\n",
    "                             )\n",
    "\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "                                 maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                                 maps_clinvar_pg4$G1['MAPS'],\n",
    "                                 maps_clinvar_pg4$G2['MAPS'],\n",
    "                                 maps_clinvar_pg4$G3['MAPS'],\n",
    "                                 maps_cgcs_signal_all_variants$cgcs[\"MAPS\"],\n",
    "                                 maps_cgcs_pg4$G1['MAPS'],\n",
    "                                 maps_cgcs_pg4$G2[\"MAPS\"],\n",
    "                                 maps_cgcs_pg4$G3[\"MAPS\"],\n",
    "                                 maps_tsgs_signal_all_variants$tsgs[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G1[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G2[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G3[\"MAPS\"]\n",
    "                               )\n",
    "\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"Alternative\", \"Alternative\", \"Alternative\", \"Alternative\",\n",
    "                             \"ClinVar\", \"ClinVar\", \"ClinVar\", \"ClinVar\",\n",
    "                             \"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\n",
    "                             \"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_gruns, function(x) length(x))\n",
    ")\n",
    "\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "\n",
    "# tiff('figures/MAPS_delta_dx.tiff', units = 'in', width = 9, height = 6, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 0.8, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Diseae Genes', size = 3)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 3)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 3)\n",
    "\n",
    "ggsave('figures/MAPS_delta_dx.tiff', units = 'in', width = 9, height = 6,device=tiff())\n",
    "ggsave('figures/MAPS_delta_dx.pdf', width = 9, height = 6, device=pdf())\n",
    "dev.off()\n",
    "\n",
    "tiff('figures/MAPS_delta_dx_resized.tiff', units = 'in', width = 6, height = 4, res = 600)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 4),\n",
    "        axis.text.y = element_text(hjust = 1, size = 7),\n",
    "        axis.title.y = element_text(size=5),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.title = element_text(size=5),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 1.3, y = 0.043, label = \"missense\", color = 'orange', size = 1.5) +\n",
    "  annotate(\"text\", x = 1.55, y = 0.005, label = \"synonymous\", color = 'grey', size = 1.5) +\n",
    "  annotate(\"text\", x = 1, y = 0.096, label = \"pLOF\", color = 'purple', size = 1.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 2) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 2)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 2)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 2)\n",
    "dev.off()\n",
    "\n",
    "# pdf('figures/MAPS_delta_dx_resized.pdf', width = 6, height = 4, res = 600)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 5),\n",
    "        axis.text.y = element_text(hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(size=9),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.title = element_text(size=5),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 1.3, y = 0.043, label = \"missense\", color = 'orange', size = 1.5) +\n",
    "  annotate(\"text\", x = 1.55, y = 0.005, label = \"synonymous\", color = 'grey', size = 1.5) +\n",
    "  annotate(\"text\", x = 1, y = 0.096, label = \"pLOF\", color = 'purple', size = 1.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 2) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 2)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 2)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor Suppressor Gene', size = 2)\n",
    "# grDevices::embedFonts('figures/MAPS_delta_dx_resized.pdf')\n",
    "ggsave('figures/MAPS_delta_dx_resized.pdf', width = 6, height = 4,device = pdf())\n",
    "ggsave('figures/MAPS_delta_dx_resized.tiff', width = 6, height = 4,device = tiff())\n",
    "dev.off()\n",
    "\n",
    "###########################################\n",
    "## Plot all G4, delta G4, and clinvar G4 ##\n",
    "###########################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'Const. SNV' = combined_promoter_const_variants_bootstrap,\n",
    "                               'Const. pG4 G1' = g4_combined_const_bootstraps$G1,\n",
    "                               'Const. pG4 G2' = g4_combined_const_bootstraps$G2,\n",
    "                               'Const. pG4 G3' = g4_combined_const_bootstraps$G3,\n",
    "\n",
    "                               'Stable SNV' = combined_promoter_stable_variants_bootstrap,\n",
    "                               'Stable pG4 G1' = g4_combined_stable_bootstraps$G1,\n",
    "                               'Stable pG4 G2' = g4_combined_stable_bootstraps$G2,\n",
    "                               'Stable pG4 G3' = g4_combined_stable_bootstraps$G3,\n",
    "\n",
    "                               'Unstable SNV' = combined_promoter_unstable_variants_bootstrap,\n",
    "                               'Unstable pG4 G1' = g4_combined_unstable_bootstraps$G1,\n",
    "                               'Unstable pG4 G2' = g4_combined_unstable_bootstraps$G2,\n",
    "                               'Unstable pG4 G3' = g4_combined_unstable_bootstraps$G3\n",
    "\n",
    ")\n",
    "\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_const$all['MAPS'],\n",
    "                                 maps_promoter_combined_const$G1['MAPS'],\n",
    "                                 maps_promoter_combined_const$G2['MAPS'],\n",
    "                                 maps_promoter_combined_const$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_stable$all['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_unstable$all['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G3['MAPS']\n",
    "\n",
    "                                 )\n",
    "\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"Alternative\", \"Alternative\", \"Alternative\", \"Alternative\",\n",
    "                             \"Const.\", \"Const.\", \"Const.\", \"Const.\",\n",
    "                             \"Stable.\", \"Stable.\", \"Stable.\", \"Stable.\",\n",
    "                             \"Unstable.\", \"Unstable.\", \"Unstable.\", \"Unstable.\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               length(const_all_promoter_variants),\n",
    "               sapply(promoter_combined_const, function(x) length(x)),\n",
    "               length(all_promoter_stable_variants),\n",
    "               sapply(promoter_combined_stable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_unstable_variants),\n",
    "               sapply(promoter_combined_unstable, function(x) length(x))\n",
    "               )\n",
    "\n",
    "plot_df$Context <- paste0(rep(c('All SNV','G4 G1', 'G4 G2', 'G4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "# tiff('figures/MAPS_delta_const.tiff', units = 'in', width = 7, height = 5, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1, y = 0.044, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 0.8, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All pG4s', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Cons. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Stable. pG4s', size = 3) +\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Unstable. pG4', size = 3)\n",
    "ggsave('figures/MAPS_delta_const.tiff', units = 'in', width = 12, height = 5, dpi = 300,device=tiff())\n",
    "ggsave('figures/MAPS_delta_const.pdf', width = 12, height = 5, device=pdf())\n",
    "dev.off()\n",
    "\n",
    "\n",
    "###########################################\n",
    "## Plot with comparison of non-pG4 Gruns ##\n",
    "###########################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'non-G4 1' = nog4_combined_bootstraps$pos1,\n",
    "                               'non-G4 2' = nog4_combined_bootstraps$pos2,\n",
    "                               'non-G4 3' = nog4_combined_bootstraps$pos3,\n",
    "                               'plof All SNV' = all_promoter_plof_bootstrap,\n",
    "                               'plof pG4 G1' = combined_g4_plof_gruns_bootstrap$G1,\n",
    "                               'plof pG4 G2' = combined_g4_plof_gruns_bootstrap$G2,\n",
    "                               'plof pG4 G3' = combined_g4_plof_gruns_bootstrap$G3,\n",
    "                               'plof non-G4 1' = combined_npg4_plof_gruns_bootstrap$pos1,\n",
    "                               'plof non-G4 2' = combined_npg4_plof_gruns_bootstrap$pos2,\n",
    "                               'plof non-G4 3' = combined_npg4_plof_gruns_bootstrap$pos3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'Delta non-G4 1' = nog4_combined_delta_bootstraps$pos1,\n",
    "                               'Delta non-G4 2' = nog4_combined_delta_bootstraps$pos2,\n",
    "                               'Delta non-G4 3' = nog4_combined_delta_bootstraps$pos3,\n",
    "                               'Const. SNV' = combined_promoter_const_variants_bootstrap,\n",
    "                               'Const. pG4 G1' = g4_combined_const_bootstraps$G1,\n",
    "                               'Const. pG4 G2' = g4_combined_const_bootstraps$G2,\n",
    "                               'Const. pG4 G3' = g4_combined_const_bootstraps$G3,\n",
    "                               'Const. non-G4 1' = nog4_combined_const_bootstraps$pos1,\n",
    "                               'Const. non-G4 2' = nog4_combined_const_bootstraps$pos2,\n",
    "                               'Const. non-G4 3' = nog4_combined_const_bootstraps$pos3,\n",
    "\n",
    "                               'Stable SNV' = combined_promoter_stable_variants_bootstrap,\n",
    "                               'Stable pG4 G1' = g4_combined_stable_bootstraps$G1,\n",
    "                               'Stable pG4 G2' = g4_combined_stable_bootstraps$G2,\n",
    "                               'Stable pG4 G3' = g4_combined_stable_bootstraps$G3,\n",
    "                               'Stable non-G4 1' = nog4_combined_stable_bootstraps$pos1,\n",
    "                               'Stable non-G4 2' = nog4_combined_stable_bootstraps$pos2,\n",
    "                               'Stable non-G4 3' = nog4_combined_stable_bootstraps$pos3,\n",
    "\n",
    "\n",
    "                               'Unstable SNV' = combined_promoter_unstable_variants_bootstrap,\n",
    "                               'Unstable pG4 G1' = g4_combined_unstable_bootstraps$G1,\n",
    "                               'Unstable pG4 G2' = g4_combined_unstable_bootstraps$G2,\n",
    "                               'Unstable pG4 G3' = g4_combined_unstable_bootstraps$G3,\n",
    "                               'Unstable non-G4 1' = nog4_combined_unstable_bootstraps$pos1,\n",
    "                               'Unstable non-G4 2' = nog4_combined_unstable_bootstraps$pos2,\n",
    "                               'Unstable non-G4 3' = nog4_combined_unstable_bootstraps$pos3,\n",
    "\n",
    "                               'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                               'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                               'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                               'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                               'ClinVar non-G4 1' = nog4_combined_clinvar_signal_bootstrap$pos1,\n",
    "                               'ClinVar non-G4 2' = nog4_combined_clinvar_signal_bootstrap$pos2,\n",
    "                               'ClinVar non-G4 3' = nog4_combined_clinvar_signal_bootstrap$pos3,\n",
    "\n",
    "                               'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                               'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                               'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                               'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "                               'Cosmic non-G4 1' = nog4_combined_cgcs_signal_bootstrap$pos1,\n",
    "                               'Cosmic non-G4 2' = nog4_combined_cgcs_signal_bootstrap$pos2,\n",
    "                               'Cosmic non-G4 3' = nog4_combined_cgcs_signal_bootstrap$pos3,\n",
    "\n",
    "                               'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                               'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                               'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                               'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3,\n",
    "                               'TsGenes non-G4 1' = nog4_combined_tsgs_signal_bootstrap$pos1,\n",
    "                               'TsGenes non-G4 2' = nog4_combined_tsgs_signal_bootstrap$pos2,\n",
    "                               'TsGenes non-G4 3' = nog4_combined_tsgs_signal_bootstrap$pos3\n",
    ")\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "                                 maps_nong4_first_last$pos1['MAPS'],\n",
    "                                 maps_nong4_first_last$pos2['MAPS'],\n",
    "                                 maps_nong4_first_last$pos3['MAPS'],\n",
    "\n",
    "                                 maps_plof_all_variants$all_promoter_plof['MAPS'],\n",
    "                                 maps_plof_gruns$G1['MAPS'],\n",
    "                                 maps_plof_gruns$G2['MAPS'],\n",
    "                                 maps_plof_gruns$G3['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos1['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos2['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_const$all['MAPS'],\n",
    "                                 maps_promoter_combined_const$G1['MAPS'],\n",
    "                                 maps_promoter_combined_const$G2['MAPS'],\n",
    "                                 maps_promoter_combined_const$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_stable$all['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_unstable$all['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos3['MAPS'],\n",
    "\n",
    "                                 maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                                 maps_clinvar_pg4$G1['MAPS'],\n",
    "                                 maps_clinvar_pg4$G2['MAPS'],\n",
    "                                 maps_clinvar_pg4$G3['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos1['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos2['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos3['MAPS'],\n",
    "\n",
    "                                 maps_cgcs_signal_all_variants$cgcs['MAPS'],\n",
    "                                 maps_cgcs_pg4$G1['MAPS'],\n",
    "                                 maps_cgcs_pg4$G2['MAPS'],\n",
    "                                 maps_cgcs_pg4$G3['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos1['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos2['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos3['MAPS'],\n",
    "\n",
    "                                 maps_tsgs_signal_all_variants$tsgs['MAPS'],\n",
    "                                 maps_tsgs_pg4$G1['MAPS'],\n",
    "                                 maps_tsgs_pg4$G2['MAPS'],\n",
    "                                 maps_tsgs_pg4$G3['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos1['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos2['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos3['MAPS']\n",
    ")\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(rep(\"All\", 7),\n",
    "                             rep('Top 1/3 LOEUF', 7),\n",
    "                             rep(\"Alternative\", 7),\n",
    "                             rep('Constitutive', 7),\n",
    "                             rep(\"Stable\", 7),\n",
    "                             rep('Unstable', 7),\n",
    "                             rep('ClinVar', 7),\n",
    "                             rep('Cancer Gene Cencus',7),\n",
    "                             rep('Tumor Suppressor Genes',7)\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_plof),\n",
    "               sapply(combined_g4_plof_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_plof, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_delta, function(x) length(x)),\n",
    "\n",
    "               length(const_all_promoter_variants),\n",
    "               sapply(promoter_combined_const, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_const, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_stable_variants),\n",
    "               sapply(promoter_combined_stable, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_stable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_unstable_variants),\n",
    "               sapply(promoter_combined_unstable, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_unstable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_clinvar_signal, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_cgcs_signal, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_tsgs_signal, function(x) length(x))\n",
    "               )\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3', 'non-G4 1', 'non-G4 2', 'non-G4 3'), 9),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4', 'non-G4', 'non-G4', 'non-G4'),9)\n",
    "\n",
    "# tiff('figures/MAPS_all_classes_3.tiff', units = 'in', width = 10, height = 4, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 32, xend = 32, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(\"MAPS\") +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\", \"blue\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 65, hjust = 1, size = 5),\n",
    "        axis.title.y = element_text(size = 8),\n",
    "        axis.text.y = element_text(size = 8),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.position = 'top',\n",
    "        legend.title = element_blank(),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 2, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.5, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 7.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 14.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 15 + 6.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 28.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 35.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 42.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 49.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 56.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 3.75, y = -0.05, label = 'All Promiters', size = 2) +\n",
    "  annotate(\"text\", x = 11.25, y = -0.05, label = 'Top 1/3 LOEUF', size = 2) +\n",
    "  annotate(\"text\", x = 17.75, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 25.25, y = -0.05, label = 'Const. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 24.25 + 7.5, y = -0.05, label = 'Stable pG4', size = 2) +\n",
    "  annotate(\"text\", x = 32.25 + 7, y = -0.05, label = 'Unstable pG4', size = 2) +\n",
    "  annotate(\"text\", x = 39.25 + 7.5, y = -0.05, label = 'ClinVar pG4', size = 2)+\n",
    "  annotate(\"text\", x = 46.25 + 7, y = -0.05, label = 'Cancer Gene Cencus pG4', size = 2)+\n",
    "  annotate(\"text\", x = 54.25 + 7, y = -0.05, label = 'Tumor Suppressor Genes pG4', size = 2)\n",
    "ggsave('figures/MAPS_all_classes_3.tiff', units = 'in', width = 10, height = 4, dpi = 300,device = tiff())\n",
    "ggsave('figures/MAPS_all_classes_3.pdf', width = 10, height = 4, device = pdf())\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42453ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 0.2kb_promoter_MAPS.R\n",
    "library(dplyr)\n",
    "library(GenomicRanges)\n",
    "library(ggplot2)\n",
    "#library(ensembldb)\n",
    "\n",
    "setwd(\"/disk2/G4_MS/frequency/MAPS_h3k4me3\")\n",
    "source(\"/disk2/G4_MS/frequency/MAPS/data/h3k4me3_stable_unstable.data.get.R\")\n",
    "load(\"/disk2/G4_MS/frequency/MAPS/data/RData/ps_lm_genomes.RData\")\n",
    "\n",
    "gtffile <- \"/disk2/G4_MS/frequency/MAPS/data/Homo_sapiens.GRCh38.104.gtf.gz\"\n",
    "DB.104 <- ensembldb::ensDbFromGtf(gtf = gtffile)\n",
    "EnsDb.Hsapiens.v104 <- ensembldb::EnsDb(DB.104)\n",
    "## cds rangeget\n",
    "mu <- data.table::fread(\"/disk2/G4_MS/frequency/MAPS/data/mutation_rate_methylation_bins.txt\")\n",
    "mu <- mu %>%\n",
    "  mutate(ref_context = paste0(context,\":\",methylation_level),\n",
    "         mut_context = paste0(substr(context,1,1),\n",
    "                              alt,\n",
    "                              substr(context,3,3)),\n",
    "         mutation = paste0(ref_context,\">\",mut_context))\n",
    "## Genes with alternativate pG4 isoform\n",
    "dlata_g4 <- unique(intersect(reference_ranges$nog4_tx$ensembl_gene_id,reference_ranges$g4_regions$ensembl_gene_id))\n",
    "constitutive_g4 <-unique(setdiff(reference_ranges$g4_regions$ensembl_gene_id,reference_ranges$nog4_tx$ensembl_gene_id))\n",
    "\n",
    "# load constraint table by gene\n",
    "gene_constraints <- data.table::fread(\"/disk2/G4_MS/reference/gnomad.v2.1.1.lof_metrics.by_gene.txt\")\n",
    "non_plof_genes <- gene_constraints[order(gene_constraints$oe_lof_upper,decreasing = TRUE),]\n",
    "non_plof_genes <- non_plof_genes[1:6568,]\n",
    "plof_genes <- gene_constraints[order(gene_constraints$oe_lof_upper,decreasing = FALSE),]\n",
    "plof_genes <- plof_genes[1:6568,]\n",
    "clinvar_genes <- data.table::fread(\"/disk2/G4_MS/reference/clinvar_20220115.vcf_annotated_af.tsv\")\n",
    "clinvar_genes <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,\n",
    "                                       keys=gsub(pattern = \":.*\",\"\",clinvar_genes$GENEINFO),\n",
    "                                       column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "cgcs <- data.table::fread(\"/disk2/G4_MS/reference/cancer_gene_census.tsv\")\n",
    "cgcs <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,keys=cgcs$GeneSymbol,column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "\n",
    "tsgs <- data.table::fread(\"/disk2/G4_MS/reference/tsgenes.tsv\")\n",
    "# tsgs <- tsgs[which(tsgs$GeneType=='protein-coding'),]\n",
    "tsgs <- AnnotationDbi::mapIds(EnsDb.Hsapiens.v104,keys=tsgs$GeneSymbol,column = \"GENEID\",keytype=\"SYMBOL\")\n",
    "\n",
    "# Load genome-wide methylation estimates\n",
    "load('/disk2/G4_MS/frequency/MAPS/data/methylation_hg38.RData')\n",
    "promoter_gnomad <- data.table::fread(\"/disk2/G4_MS/frequency/MAPS/data/active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.tsv\")\n",
    "colnames(promoter_gnomad) <- c(\"CHROM\", \"POS\", \"REF\", \"ALT\", \"AN\", \"AC\", \"AF\")\n",
    "promoter_hept <- data.table::fread(\"/disk2/G4_MS/frequency/MAPS/data/active_promoter_region_gnomad.genomes.v3.1.sites.pass_snps.seq\", header = FALSE)\n",
    "promoter_tri <- substr(promoter_hept$V2,3,5)\n",
    "promoter_gnomad$context <- toupper(promoter_tri)\n",
    "promoter_gnomad$mut_context <- paste0(substr(promoter_gnomad$context,1,1),\n",
    "                                      promoter_gnomad$ALT,\n",
    "                                      substr(promoter_gnomad$context,3,3))\n",
    "promoter_gnomad$hept <- toupper(promoter_hept$V2)\n",
    "promoter_gnomad <- promoter_gnomad %>%  select(CHROM, POS, REF, ALT, AN, AC, AF, context, mut_context, hept) %>%\n",
    "  filter(CHROM != 'chrX' & CHROM != 'chrY', AN > 0.8*152312, AC >= 1)\n",
    "rm(promoter_hept)\n",
    "rm(promoter_tri)\n",
    "promoter_gnomad_grange <- granges_from_vcf(promoter_gnomad)\n",
    "### \n",
    "# combined_promoter_variants = promoter_gnomad_grange\n",
    "g1_allowed <- c('AGG',\n",
    "                'CCA',\n",
    "                'CCG',\n",
    "                'CCT',\n",
    "                'CGG',\n",
    "                'TGG')\n",
    "g2_allowed <- c('CCC','GGG')\n",
    "g3_allowed <- c('ACC',\n",
    "                'GCC',\n",
    "                'GGA',\n",
    "                'GGC',\n",
    "                'GGT',\n",
    "                'TCC')\n",
    "\n",
    "allowed <- list('pos1' = g1_allowed,\n",
    "                'pos2' = g2_allowed,\n",
    "                'pos3' = g3_allowed)\n",
    "\n",
    "######################\n",
    "## Helper Functions ##\n",
    "######################\n",
    "\n",
    "get_ps <- function(vcf_list, mu) {\n",
    "\n",
    "  n <- lapply(vcf_list, function(x) table(x$mutation))\n",
    "  ps <- lapply(vcf_list, function(x) table(x[x$AC == 1]$mutation))\n",
    "\n",
    "  ps_res <- mapply(function(x, y) cbind(x, y[match(names(x), names(y))]), x=n, y=ps, SIMPLIFY = FALSE)\n",
    "  ps_res <- lapply(ps_res, function(x) x[match(mu$mutation, rownames(x)),])\n",
    "\n",
    "  return(ps_res)\n",
    "}\n",
    "\n",
    "\n",
    "get_mu_avg <- function(ps_list, mu) {\n",
    "\n",
    "  res <- lapply(ps_list, function(x) sum((x*mu$mu_snp), na.rm = TRUE) / sum(x, na.rm = TRUE))\n",
    "\n",
    "\n",
    "  return(res)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "calc_maps <- function(raw_ps, ps_counts, predictions) {\n",
    "\n",
    "  obs <- unlist(raw_ps) * unlist(ps_counts)\n",
    "  exp <- unlist(predictions) * unlist(ps_counts)\n",
    "\n",
    "  maps <- (obs - exp) / unlist(ps_counts)\n",
    "\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "# MAPS 2 model:\n",
    "calc_maps2 <- function(variants, mu, ps_lm) {\n",
    "\n",
    "  variants <- data.frame(variants)\n",
    "\n",
    "  total_variants <- dim(variants)[[1]]\n",
    "\n",
    "  variants <- split(variants, variants$group)\n",
    "\n",
    "  if (length(variants) == 2) {\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else if (names(variants) == 'no_adjust') {\n",
    "\n",
    "    methyl_adjust <- c('exp_singletons.1' = 0,\n",
    "                       'obs_singletons' = 0)\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "  } else {\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "    no_adjust <- c('exp_singletons.1' = 0,\n",
    "                   'obs_singletons' = 0)\n",
    "  }\n",
    "  combined <- methyl_adjust + no_adjust\n",
    "\n",
    "  maps <- (combined['obs_singletons'] - combined['exp_singletons.1']) / total_variants\n",
    "  raw_p <- combined['obs_singletons'] / total_variants\n",
    "  se <-  sqrt(raw_p*(1-raw_p)/total_variants)\n",
    "  result <- c(\"MAPS\" = as.numeric(maps), \"SE\" = as.numeric(se), \"N\" = as.integer(total_variants))\n",
    "  return(result)\n",
    "\n",
    "}\n",
    "\n",
    "get_obs_exp_singletons <- function(variants, ps_lm) {\n",
    "\n",
    "  variants <- na.omit(variants)\n",
    "  mu_avg <- sum(variants$n * variants$mu_snp) / sum(variants$n)\n",
    "  exp_singletons <- sum(variants$n) * predict(ps_lm, data.frame('mu_snp' = mu_avg))\n",
    "  obs_singletons <- sum(variants$singletons)\n",
    "  return(c('exp_singletons' = exp_singletons,\n",
    "           'obs_singletons' = obs_singletons))\n",
    "\n",
    "}\n",
    "\n",
    "do_MAPS <- function(vcf_list, mu, ps_model) {\n",
    "  # determine proportion singletons\n",
    "  ps <- get_ps(vcf_list, mu)\n",
    "  # determine mu average for the variants + contexts\n",
    "  mu_avg <- get_mu_avg(ps, mu)\n",
    "  # extract raw proportion of singletons\n",
    "  raw_p <- lapply(vcf_list, function(x) length(x[x$AC==1])/length(x))\n",
    "  var_counts <- lapply(vcf_list, function(x) length(x))\n",
    "  se <- mapply(function(p, n ) sqrt(p*(1-p)/n), raw_p, var_counts)\n",
    "  # predicted ps:\n",
    "  ps_predict <- lapply(mu_avg, function(x) predict(ps_model, data.frame(mu = unlist(x))))\n",
    "  maps <- calc_maps(raw_p, var_counts, ps_predict)\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "\n",
    "annotate_grun_position <- function(grun_grange, promoter_region, width = 3) {\n",
    "\n",
    "  gruns <- grun_grange[width(grun_grange) == width]\n",
    "  gruns_df <- data.frame(gruns)\n",
    "  #gruns$strand <- gruns$oristrand\n",
    "  gruns_df$strand <- promoter_region$oristrand[match(gruns_df$txid, promoter_region$txid)]\n",
    "  # gruns_df$strand <- gruns_df$strand.1\n",
    "  # gruns_df <- gruns_df[,!colnames(gruns_df) %in% c(\"oristrand\",\"strand.1\")]\n",
    "  tile <- makeGRangesFromDataFrame(gruns_df)\n",
    "  tile <- tile(tile, width = 1) %>% as.data.frame()\n",
    "  pos_tiles <- tile %>% filter(strand == '+')\n",
    "  neg_tiles <- tile %>% filter(strand == '-')\n",
    "\n",
    "  pos_tiles$var_pos <- rep(x = c(1:width), dim(pos_tiles)[[1]]/width)\n",
    "  neg_tiles$var_pos <- rep(x = c(1:width), dim(neg_tiles)[[1]]/width)\n",
    "\n",
    "  tiles <- rbind(pos_tiles, neg_tiles)\n",
    "\n",
    "  pos1 <- tiles %>% filter(var_pos == 1)\n",
    "  pos2 <- tiles %>% filter(var_pos == 2)\n",
    "  pos3 <- tiles %>% filter(var_pos == 3)\n",
    "\n",
    "  res <- list(\"pos1\" = pos1,\n",
    "              \"pos2\" = pos2,\n",
    "              \"pos3\" = pos3)\n",
    "  res <- lapply(res, function(x) makeGRangesFromDataFrame(x, keep.extra.columns = TRUE))\n",
    "  return(res)\n",
    "\n",
    "}\n",
    "match_var_to_strand <- function(var_grange_first, var_grange_last, tx_grange) {\n",
    "\n",
    "  # this function annotates a variant as being either first or last in a trinucleotide G/C run\n",
    "  # based on the strand of the UTR transcript\n",
    "\n",
    "  pos_strand <- tx_grange[tx_grange$oristrand == '+']\n",
    "  neg_strand <- tx_grange[tx_grange$oristrand == '-']\n",
    "  pos_strand_first <- subsetByOverlaps(var_grange_first, pos_strand)\n",
    "  pos_strand_last <- subsetByOverlaps(var_grange_last, pos_strand)\n",
    "\n",
    "  neg_strand_first <- subsetByOverlaps(var_grange_last, neg_strand)\n",
    "  neg_strand_last <- subsetByOverlaps(var_grange_first, neg_strand)\n",
    "\n",
    "  first_vars <- c(pos_strand_first, neg_strand_first)\n",
    "  last_vars <- c(pos_strand_last, neg_strand_last)\n",
    "\n",
    "  return(list('pos1'=first_vars,\n",
    "              'pos3'=last_vars))\n",
    "\n",
    "}\n",
    "\n",
    "calc_maps2_bootstrap <- function(variants, mu, ps_lm) {\n",
    "\n",
    "  total_variants <- dim(variants)[[1]]\n",
    "\n",
    "  variants <- variants[sample(1:total_variants, total_variants, replace = TRUE),]\n",
    "\n",
    "  variants <- split(variants, variants$group)\n",
    "\n",
    "  if (length(variants) == 2) {\n",
    "\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else if (names(variants) == 'no_adjust') {\n",
    "\n",
    "    methyl_adjust <- c('exp_singletons.1' = 0,\n",
    "                       'obs_singletons' = 0)\n",
    "\n",
    "    no_adjust <- variants$no_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$no_adjust)\n",
    "\n",
    "  } else {\n",
    "\n",
    "    methyl_adjust <- variants$methyl_adjust %>%\n",
    "      group_by(mut) %>%\n",
    "      summarise(singletons = length(which(AC == 1)),\n",
    "                n = length(AN)) %>%\n",
    "      mutate(ps = singletons / n) %>%\n",
    "      left_join(., mu, by = c('mut' = 'mutation')) %>%\n",
    "      select(mut, singletons, n, ps, methylation_level, mu_snp) %>%\n",
    "      get_obs_exp_singletons(., ps_lm = ps_lm$methyl_adjust)\n",
    "\n",
    "    no_adjust <- c('exp_singletons.1' = 0,\n",
    "                   'obs_singletons' = 0)\n",
    "  }\n",
    "combined <- methyl_adjust + no_adjust\n",
    "\n",
    "  maps <- (combined['obs_singletons'] - combined['exp_singletons.1']) / total_variants\n",
    "\n",
    "  return(as.numeric(maps))\n",
    "\n",
    "}\n",
    "                \n",
    "bootstrap_maps2 <- function(variants, mu, ps_lm, ntimes = 10000) {\n",
    "\n",
    "  variants <- data.frame(variants)\n",
    "\n",
    "  variants <- variants %>%\n",
    "    select(AC, AN, group, mut)\n",
    "\n",
    "  res <- vector(\"list\", length = ntimes)\n",
    "\n",
    "  res <- lapply(res, function(x) calc_maps2_bootstrap(variants, mu, ps_lm))\n",
    "\n",
    "  return(unlist(res))\n",
    "\n",
    "}\n",
    "\n",
    "resample_MAPS <- function(vcf, mu, ps_model) {\n",
    "\n",
    "  n <- table(vcf$mut_context)\n",
    "  ps <- table(vcf[vcf$AC == 1]$mut_context)\n",
    "\n",
    "  ps <- cbind(n, ps[match(names(n), names(ps))])\n",
    "  ps <- ps[match(mu$mut_context, rownames(ps)),]\n",
    "\n",
    "  mu_average <- sum(ps*mu$mu_snp, na.rm =TRUE) / sum(ps, na.rm = TRUE)\n",
    "\n",
    "  raw_p <- length(vcf[vcf$AC==1])/length(vcf)\n",
    "  var_counts <- length(vcf)\n",
    "  ps_predict <- predict(ps_model, data.frame(mu = unlist(mu_average)))\n",
    "\n",
    "  obs <- unlist(raw_p) * unlist(var_counts)\n",
    "  exp <- unlist(ps_predict) * unlist(var_counts)\n",
    "  maps <- (obs-exp)/unlist(var_counts)\n",
    "\n",
    "  return(maps)\n",
    "\n",
    "}\n",
    "\n",
    "generate_bootstrap <- function(vcf, n, mu, ps_model) {\n",
    "\n",
    "  x <- length(vcf)\n",
    "\n",
    "  res <- rep(x, n)\n",
    "\n",
    "  samples <- lapply(res, function(x) sample(x, size = x, replace = TRUE))\n",
    "\n",
    "  maps_bootstrap <- lapply(samples, function(x) resample_MAPS(vcf[x], mu = mu, ps_model = ps_model))\n",
    "\n",
    "  return(unlist(maps_bootstrap))\n",
    "\n",
    "}\n",
    "                \n",
    "chunked_bootstrap <- function(vcf, n, mu, ps_model) {\n",
    "\n",
    "  chunksize <- as.integer(n/10)\n",
    "  chunk1 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk2 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk3 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk4 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk5 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk6 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk7 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk8 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk9 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                            mu = mu, ps_lm = ps_model)\n",
    "  chunk10 <- bootstrap_maps2(variants = vcf, ntimes = chunksize,\n",
    "                             mu = mu, ps_lm = ps_model)\n",
    "  result <- c(chunk1, chunk2, chunk3, chunk4, chunk5,\n",
    "              chunk6, chunk7, chunk8, chunk9, chunk10)\n",
    "  return(result)\n",
    "\n",
    "}\n",
    "annotate_methylation <- function(variants, mutation_table, methylation) {\n",
    "\n",
    "  variants <- left_join(variants, methylation)\n",
    "  # set missing methylation values to zero\n",
    "  variants$methyl_bins[is.na(variants$methyl_bins)] <- 0\n",
    "  fixed_context <- ifelse(variants$context %in% mutation_table$context,\n",
    "                          variants$context,\n",
    "                          chartr('ATGC','TACG',variants$context))\n",
    "  fixed_mut <- ifelse(variants$context %in% mutation_table$context,\n",
    "                      variants$mut_context,\n",
    "                      chartr('ATGC','TACG',variants$mut_context))\n",
    "\n",
    "  # amend mutation context for MAPS score calculation\n",
    "  variants$mut <- paste0(fixed_context,':',\n",
    "                         variants$methyl_bins,\n",
    "                         '>',\n",
    "                         fixed_mut)\n",
    "  variants$mut <- ifelse(variants$mut %in% mutation_table$mutation,\n",
    "                         variants$mut,\n",
    "                         gsub('[1|2]',\n",
    "                              '0',\n",
    "                              variants$mut))\n",
    "  variants$context <- fixed_context\n",
    "  variants$mut_context <- fixed_mut\n",
    "\n",
    "  # annotate variants for methylation adjustment\n",
    "  var_adjust <- mutation_table %>%\n",
    "    filter(methylation_level != 0) %>%\n",
    "    select(context) %>% unlist()\n",
    "  variants$group <- ifelse((variants$context %in% var_adjust),\n",
    "                           ifelse((variants$REF == 'C' & variants$ALT == 'T') | (variants$REF == 'G' & variants$ALT == 'A'),\n",
    "                                  'methyl_adjust',\n",
    "                                  'no_adjust'),\n",
    "                           'no_adjust')\n",
    "\n",
    "  return(variants)\n",
    "\n",
    "}\n",
    "\n",
    "##############\n",
    "## Analysis ##\n",
    "##############\n",
    "# annotate methylation context for MAPS analysis\n",
    "# correct gruns\n",
    "reference_granges$g4_gruns <- subsetByOverlaps(reference_granges$g4_gruns, promoter_region)\n",
    "## combined_promoter_variants = promoter_gnomad_grange\n",
    "promoter_gnomad_grange_df <- data.frame(promoter_gnomad_grange) %>% mutate(CHROM = seqnames, POS = start)\n",
    "promoter_gnomad_grange_df <- annotate_methylation(data.frame(promoter_gnomad_grange_df),\n",
    "                                                  mu,\n",
    "                                                  combined_methylation_estimates)\n",
    "promoter_gnomad_grange <- promoter_gnomad_grange_df %>% select(-c(CHROM, POS)) %>%\n",
    "  makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "\n",
    "# subset pG4s by alternative vs. constitutive\n",
    "promoter_delta_pg4 <- reference_granges$g4_gruns[reference_granges$g4_gruns$ensembl_gene_id %in% dlata_g4]\n",
    "promoter_const_pg4 <- reference_granges$g4_gruns[!reference_granges$g4_gruns$ensembl_gene_id %in% dlata_g4]\n",
    "                           \n",
    "promoter_stable_pg4 <- reference_granges$stable_g4_gruns\n",
    "# promoter_stable_pg4 <- promoter_stable_pg4[!promoter_stable_pg4$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "# promoter_unstable_pg4 <- reference_granges$g4_gruns[reference_granges$g4_gruns$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "promoter_unstable_pg4 <- reference_granges$unstable_g4_gruns\n",
    "# Annotate grun positions (G1, G2, G3) and remove sequence contexts that are mismatched\n",
    "promoter_g4_gruns_positional <- annotate_grun_position(reference_granges$g4_gruns,promoter_region)\n",
    "promoter_g4_vars <- lapply(promoter_g4_gruns_positional, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for delta pG4\n",
    "promoter_delta_pg4 <- annotate_grun_position(promoter_delta_pg4,promoter_region)\n",
    "promoter_delta_pg4 <- lapply(promoter_delta_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for constitutive pG4\n",
    "promoter_const_pg4 <- annotate_grun_position(promoter_const_pg4,promoter_region)\n",
    "promoter_const_pg4 <- lapply(promoter_const_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))                     \n",
    "\n",
    "# repeat for stable pG4\n",
    "promoter_stable_pg4 <- annotate_grun_position(promoter_stable_pg4,promoter_region)\n",
    "promoter_stable_pg4 <- lapply(promoter_stable_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# repeat for unstable pG4\n",
    "promoter_unstable_pg4 <- annotate_grun_position(promoter_unstable_pg4,promoter_region)\n",
    "promoter_unstable_pg4 <- lapply(promoter_unstable_pg4, function(x) subsetByOverlaps(promoter_gnomad_grange, x))\n",
    "\n",
    "# combine variant sets and perform MAPS\n",
    "promoter_combined <- c('G1' = promoter_g4_vars$pos1,\n",
    "                  'G2' = promoter_g4_vars$pos2,\n",
    "                  'G3' = promoter_g4_vars$pos3\n",
    ")\n",
    "promoter_combined_delta <- c('G1' = promoter_delta_pg4$pos1,\n",
    "                             'G2' = promoter_delta_pg4$pos2,\n",
    "                             'G3' = promoter_delta_pg4$pos3\n",
    ")\n",
    "promoter_combined_const <- c('G1' = promoter_const_pg4$pos1,\n",
    "                             'G2' = promoter_const_pg4$pos2,\n",
    "                             'G3' = promoter_const_pg4$pos3\n",
    ")\n",
    "###\n",
    "promoter_combined_stable <- c('G1' = promoter_stable_pg4$pos1,\n",
    "                             'G2' = promoter_stable_pg4$pos2,\n",
    "                             'G3' = promoter_stable_pg4$pos3\n",
    ")\n",
    "promoter_combined_unstable <- c('G1' = promoter_unstable_pg4$pos1,\n",
    "                             'G2' = promoter_unstable_pg4$pos2,\n",
    "                             'G3' = promoter_unstable_pg4$pos3\n",
    ")\n",
    "# Evaluate MAPS\n",
    "maps_pg4_combined <- lapply(promoter_combined, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_delta <- lapply(promoter_combined_delta, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_const<- lapply(promoter_combined_const, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_stable <- lapply(promoter_combined_stable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_promoter_combined_unstable <- lapply(promoter_combined_unstable, function(x) calc_maps2(x, mu, ps_lm_genomes))                                       \n",
    "\n",
    "######################################################################\n",
    "## repeat for GGG/CCC repeats to control for local sequence context ##\n",
    "######################################################################\n",
    "\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_gnomad_grange, promoter_combined$G1, invert = TRUE)\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, promoter_combined$G2, invert = TRUE)\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, promoter_combined$G3, invert = TRUE)\n",
    "# remove CDS\n",
    "promoter_combined_npg4 <- subsetByOverlaps(promoter_combined_npg4, cds_ranges, invert = TRUE)\n",
    "# remove CpG sites\n",
    "promoter_combined_npg4_central <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 2, 6))\n",
    "promoter_combined_npg4_central <- promoter_combined_npg4[promoter_combined_npg4_central]\n",
    "\n",
    "# Map 1st and 3rd positions for non-pG4 sequences:\n",
    "promoter_region_df <- data.frame(promoter_region)\n",
    "promoter_region_df$strand <- promoter_region_df$oristrand\n",
    "# promoter_region_df <- promoter_region_df[,!colnames(promoter_region_df) %in% c(\"oristrand\")]\n",
    "promoter_region_df <- makeGRangesFromDataFrame(promoter_region_df,\n",
    "                                               keep.extra.columns = TRUE,\n",
    "                                               seqnames.field = 'seqnames',\n",
    "                                               start.field = 'start',\n",
    "                                               end.field = 'end')\n",
    "\n",
    "promoter_combined_npg4_first <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 3, 7))\n",
    "promoter_combined_npg4_first <- promoter_combined_npg4[promoter_combined_npg4_first]\n",
    "\n",
    "promoter_combined_npg4_last <- grepl('[ATG]CCC[ATG]|[ATC]GGG[ATC]', substr(promoter_combined_npg4$hept, 1, 5))\n",
    "promoter_combined_npg4_last <- promoter_combined_npg4[promoter_combined_npg4_last]\n",
    "\n",
    "first_last_npg4 <- match_var_to_strand(promoter_combined_npg4_first, promoter_combined_npg4_last,\n",
    "                                       promoter_region_df)\n",
    "\n",
    "promoter_nog4_combined <- list('pos1' = first_last_npg4$pos1,\n",
    "                      'pos2' = promoter_combined_npg4_central,\n",
    "                      'pos3' = first_last_npg4$pos3)\n",
    "# All non-pG4 GGG / CCC contexts\n",
    "maps_nong4_first_last <- lapply(promoter_nog4_combined, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_all_promoter <- list('all'=calc_maps2(promoter_gnomad_grange, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 delta GGG /CCC contexts\n",
    "promoter_nog4_combined_delta <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_delta <- lapply(promoter_nog4_combined_delta, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_delta_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    ")\n",
    "all_promoter_delta_variants <- subsetByOverlaps(all_promoter_delta_variants,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    "                                                )\n",
    "maps_all_promoter_delta <- list('all'=calc_maps2(all_promoter_delta_variants, mu, ps_lm_genomes))\n",
    "                                \n",
    "# non-pG4 const GGG /CCC contexts\n",
    "promoter_nog4_combined_const <- lapply(promoter_nog4_combined, function(x)\n",
    "  subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% dlata_g4]\n",
    "))\n",
    "promoter_nog4_combined_const <- lapply(promoter_nog4_combined_const, function(x)\n",
    "  subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    "  )\n",
    ")\n",
    "\n",
    "maps_promoter_nog4_const <- lapply(promoter_nog4_combined_const, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "# select variants that are not in delta pG4 genes' promoter\n",
    "const_all_promoter_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% dlata_g4])\n",
    "# subset these variants by those that are in pG4 genes to isolate \"constitutive\" pG4s\n",
    "const_all_promoter_variants <- subsetByOverlaps(const_all_promoter_variants,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$g4_regions$ensembl_gene_id]\n",
    ")\n",
    "maps_all_promoter_const <- list('all'=calc_maps2(const_all_promoter_variants, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 stable GGG/CCC contexts\n",
    "promoter_nog4_combined_stable <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "promoter_nog4_combined_stable <- lapply(promoter_nog4_combined_stable, function(x) subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_stable <- lapply(promoter_nog4_combined_stable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_stable_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    ")\n",
    "all_promoter_stable_variants <- subsetByOverlaps(all_promoter_stable_variants,\n",
    "                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    ")\n",
    "maps_all_promoter_stable <- list('all'=calc_maps2(all_promoter_stable_variants, mu, ps_lm_genomes))\n",
    "\n",
    "# non-pG4 unstable GGG/CCC contexts\n",
    "promoter_nog4_combined_unstable <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x, promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "promoter_nog4_combined_unstable <- lapply(promoter_nog4_combined_unstable, function(x) subsetByOverlaps(x, promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    "                                                                          ))\n",
    "maps_promoter_nog4_unstable <- lapply(promoter_nog4_combined_unstable, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "all_promoter_unstable_variants <- subsetByOverlaps(promoter_gnomad_grange,\n",
    "                                           promoter_region_df[promoter_region_df$ensembl_gene_id %in% reference_granges$unstable_g4_gruns$ensembl_gene_id]\n",
    ")\n",
    "all_promoter_unstable_variants <- subsetByOverlaps(all_promoter_unstable_variants,\n",
    "                                           promoter_region_df[!promoter_region_df$ensembl_gene_id %in% reference_granges$stable_g4_gruns$ensembl_gene_id]\n",
    ")\n",
    "maps_all_promoter_unstable <- list('all'=calc_maps2(all_promoter_unstable_variants, mu, ps_lm_genomes)) \n",
    "                                      \n",
    "####################################################################\n",
    "## Evaluate MAPS for variants in top 1/3 of most constrained genes ##\n",
    "#####################################################################\n",
    "## promoter_g4_gruns = combined_g4_gruns\n",
    "promoter_g4_gruns <- reference_granges$g4_gruns[width(reference_granges$g4_gruns) == 3]\n",
    "promoter_g4_gruns$plof <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% plof_genes$gene_id,\n",
    "                                 TRUE,\n",
    "                                 FALSE)\n",
    "\n",
    "promoter_combined_g4_plof_gruns <- lapply(promoter_combined,\n",
    "                                 function(x) subsetByOverlaps(\n",
    "                                   x, promoter_g4_gruns[promoter_g4_gruns$plof == TRUE])\n",
    "                                 )\n",
    "\n",
    "# overlap gnomAD variants\n",
    "combined_g4_plof_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$plof == TRUE]))\n",
    "\n",
    "#combined_g4_nonplof_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$non_plof == TRUE]))\n",
    "\n",
    "# evaluate MAPS for top third most constrained UTR\n",
    "maps_plof_gruns <-  lapply(combined_g4_plof_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "# Evaluate all variants in top 1/3 most constrained UTR\n",
    "promoter_reference_grange <- data.frame(promoter_region) %>%\n",
    "  select(txid,\n",
    "         ensembl_gene_id,\n",
    "         start,\n",
    "         end,\n",
    "         oristrand,\n",
    "         seqnames)\n",
    "promoter_reference_grange$strand <- promoter_reference_grange$oristrand\n",
    "promoter_reference_grange$top_third <- ifelse(promoter_reference_grange$ensembl_gene_id %in% plof_genes$gene_id, TRUE, FALSE)\n",
    "promoter_reference_grange$bottom_third <- ifelse(promoter_reference_grange$ensembl_gene_id %in% non_plof_genes$gene_id, TRUE,FALSE)\n",
    "promoter_reference_grange$clinvar_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% clinvar_genes, TRUE, FALSE)\n",
    "promoter_reference_grange$cgcs_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% cgcs, TRUE, FALSE)\n",
    "promoter_reference_grange$tsgs_signal <- ifelse(promoter_reference_grange$ensembl_gene_id %in% tsgs, TRUE, FALSE)\n",
    "promoter_reference_grange <- promoter_reference_grange %>% makeGRangesFromDataFrame(start.field = 'start',\n",
    "                                                                            end.field = 'end',\n",
    "                                                                            seqnames.field = 'seqnames', keep.extra.columns =TRUE)\n",
    "\n",
    "# define variant subsets\n",
    "all_promoter_plof <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$top_third == TRUE])\n",
    "all_promoter_nonplof <- subsetByOverlaps(promoter_gnomad_grange,promoter_reference_grange[promoter_reference_grange$bottom_third == TRUE])\n",
    "all_promoter_clinvar_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$clinvar_signal == TRUE])\n",
    "all_promoter_cgcs_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$cgcs_signal == TRUE])\n",
    "all_promoter_tsgs_signal <- subsetByOverlaps(promoter_gnomad_grange, promoter_reference_grange[promoter_reference_grange$tsgs_signal == TRUE])\n",
    "\n",
    "nog4_combined_plof <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$top_third == TRUE]))\n",
    "nog4_combined_clinvar_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$clinvar_signal == TRUE]))\n",
    "nog4_combined_cgcs_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$cgcs_signal == TRUE]))\n",
    "nog4_combined_tsgs_signal <- lapply(promoter_nog4_combined, function(x) subsetByOverlaps(x,promoter_reference_grange[promoter_reference_grange$tsgs_signal == TRUE]))\n",
    "\n",
    "\n",
    "promoter_g4_gruns$clinvar_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% clinvar_genes, TRUE, FALSE)\n",
    "promoter_g4_gruns$cgcs_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% cgcs, TRUE, FALSE)\n",
    "promoter_g4_gruns$tsgs_signal <- ifelse(promoter_g4_gruns$ensembl_gene_id %in% tsgs, TRUE, FALSE)\n",
    "\n",
    "#####################################\n",
    "## evaluate MAPS for ClinVar genes ##\n",
    "#####################################\n",
    "maps_plof_all_variants <- list('all_promoter_plof' = calc_maps2(all_promoter_plof, mu, ps_lm_genomes))\n",
    "\n",
    "maps_nog4_combined_plof <- lapply(nog4_combined_plof, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "maps_clinvar_signal_all_variants <- list('clinvar' = calc_maps2(all_promoter_clinvar_signal, mu, ps_lm_genomes))\n",
    "maps_cgcs_signal_all_variants <- list('cgcs' = calc_maps2(all_promoter_cgcs_signal, mu, ps_lm_genomes))\n",
    "maps_tsgs_signal_all_variants <- list('tsgs' = calc_maps2(all_promoter_tsgs_signal, mu, ps_lm_genomes))\n",
    "\n",
    "combined_g4_clinvar_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$clinvar_signal == TRUE]))\n",
    "combined_g4_cgcs_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$cgcs_signal == TRUE]))\n",
    "combined_g4_tsgs_signal_gruns <- lapply(promoter_combined, function(x) subsetByOverlaps(x, promoter_g4_gruns[promoter_g4_gruns$tsgs_signal == TRUE]))\n",
    "\n",
    "maps_clinvar_pg4 <- lapply(combined_g4_clinvar_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_pg4 <- lapply(combined_g4_cgcs_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_pg4 <- lapply(combined_g4_tsgs_signal_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "maps_clinvar_nong4 <- lapply(nog4_combined_clinvar_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_nong4 <- lapply(nog4_combined_cgcs_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_nong4 <- lapply(nog4_combined_tsgs_signal, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "\n",
    "combined_g4_clinvar_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_clinvar_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "combined_g4_cgcs_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_cgcs_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "combined_g4_tsgs_signal_delta_gruns <- mapply(function(x,y) subsetByOverlaps(x,y),\n",
    "                                         x = combined_g4_tsgs_signal_gruns,\n",
    "                                         y = promoter_combined_delta)\n",
    "\n",
    "\n",
    "maps_clinvar_pg4_delta <- lapply(combined_g4_clinvar_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_cgcs_pg4_delta <- lapply(combined_g4_cgcs_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "maps_tsgs_pg4_delta <- lapply(combined_g4_tsgs_signal_delta_gruns, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "\n",
    "# Generate MAPS estimates for missense and putative loss of function-causing variants in gnomAD\n",
    "missense <- data.table::fread('/disk2/G4_MS/frequency/MAPS/data/gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.tsv',header=FALSE)\n",
    "colnames(missense) <- c('CHROM','POS','REF','ALT','AN','AC','AF','type','other', 'vep')\n",
    "heptamers <- data.table::fread(\"/disk2/G4_MS/frequency/MAPS/data/gnomad.genomes.v3.1.sites.missense.pass_snps.filtered.seq\", header = FALSE)\n",
    "missense$heptamers <- toupper(heptamers$V2)\n",
    "missense$context <- toupper(substr(heptamers$V2, 3, 5))\n",
    "\n",
    "rm(heptamers)\n",
    "                              \n",
    "an_adj <- 0.8*152312\n",
    "missense <- missense %>% filter(AC != 0,\n",
    "                                AN >= an_adj, CHROM != \"chrX\" & CHROM != \"chrY\") %>%\n",
    "                        select(-c(type, other, vep))\n",
    "\n",
    "missense$mut_context <- paste0(substr(missense$context, 1, 1),\n",
    "                               missense$ALT,\n",
    "                               substr(missense$context, 3, 3))\n",
    "\n",
    "missense$CHROM <- as.character(missense$CHROM)\n",
    "missense <- annotate_methylation(missense, mu, combined_methylation_estimates)\n",
    "plof <- data.table::fread(\"/disk2/G4_MS/frequency/MAPS/data/gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv\",\n",
    "                          header=FALSE,\n",
    "                          col.names = c('CHROM','POS','REF','ALT','AN','AC','AF','type','other', 'vep')\n",
    ")\n",
    "heptamers <- data.table::fread('/disk2/G4_MS/frequency/MAPS/data/gnomad.genomes.v3.1.sites.pass_snps.high_plof.tsv.seq', header = FALSE)\n",
    "plof$heptamers <- heptamers$V2 %>% toupper()\n",
    "rm(heptamers)\n",
    "plof <- plof %>% filter(AC >= 1,\n",
    "                        AN >= 0.8*152312, CHROM != \"chrX\" & CHROM != \"chrY\",\n",
    "                        nchar(REF) == 1, nchar(ALT) == 1) %>%\n",
    "  select(-c(type, other, vep))\n",
    "\n",
    "plof$context <- substr(plof$heptamers, 3,5)\n",
    "\n",
    "plof$mut_context <- paste0(substr(plof$context, 1, 1),\n",
    "                           plof$ALT,\n",
    "                           substr(plof$context, 3, 3))\n",
    "\n",
    "plof$CHROM <- as.character(plof$CHROM)\n",
    "plof <- annotate_methylation(plof, mu, combined_methylation_estimates)\n",
    "####\n",
    "cds_varsets <- list('missense_all' = missense,\n",
    "                    'lof_all' = plof)\n",
    "maps_cds_markers <- lapply(cds_varsets, function(x) calc_maps2(x, mu, ps_lm_genomes))\n",
    "##############################\n",
    "## Generate bootstrapped CI ##\n",
    "##############################\n",
    "# G4_combined bootstraps\n",
    "g4_combined_bootstraps <- lapply(promoter_combined, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "save(g4_combined_bootstraps, file = 'RData/g4_positional_maps_bootstraps_ncpg.RData')\n",
    "lapply(g4_combined_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/g4_positional_maps_bootstraps_ncpg.RData')\n",
    "\n",
    "# # G4 delta bootstraps\n",
    "g4_combined_delta_bootstraps <- lapply(promoter_combined_delta, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000)) #lapply(utr_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(g4_combined_delta_bootstraps, file = 'RData/g4_combined_delta_bootstraps.RData')\n",
    "lapply(g4_combined_delta_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/g4_combined_delta_bootstraps.RData\")\n",
    "\n",
    "# # G4 const bootstraps\n",
    "g4_combined_const_bootstraps <- lapply(promoter_combined_const, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(utr_combined_const, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(g4_combined_const_bootstraps, file = 'RData/g4_combined_const_bootstraps.RData')\n",
    "lapply(g4_combined_const_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/g4_combined_const_bootstraps.RData\")\n",
    "       \n",
    "# G4 stable bootstraps\n",
    "#g4_combined_stable_bootstraps <- lapply(promoter_combined_stable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "#save(g4_combined_stable_bootstraps, file = 'RData/g4_combined_stable_bootstraps.RData')\n",
    "#lapply(g4_combined_stable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "load(\"RData/g4_combined_stable_bootstraps.RData\")\n",
    "\n",
    "# G4 unstable bootstraps\n",
    "g4_combined_unstable_bootstraps <- lapply(promoter_combined_unstable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "save(g4_combined_unstable_bootstraps, file = 'RData/g4_combined_unstable_bootstraps.RData')\n",
    "lapply(g4_combined_unstable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/g4_combined_unstable_bootstraps.RData\")\n",
    "\n",
    "# # non-G4 G bootstraps\n",
    "nog4_combined_bootstraps <- lapply(promoter_nog4_combined, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_bootstraps, file = 'RData/nog4_combined_bootstraps.RData')\n",
    "lapply(nog4_combined_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/nog4_combined_bootstraps.RData\")\n",
    "\n",
    "# # non-G4 delta-G bootstraps\n",
    "nog4_combined_delta_bootstraps <-  lapply(promoter_nog4_combined_delta, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_delta_bootstraps, file = 'RData/nog4_combined_delta_bootstraps.RData')\n",
    "lapply(nog4_combined_delta_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_delta_bootstraps.RData')\n",
    "\n",
    "# # non-G4 const-G bootstraps\n",
    "nog4_combined_const_bootstraps <-  lapply(promoter_nog4_combined_const, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_const, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_const_bootstraps, file = 'RData/nog4_combined_const_bootstraps.RData')\n",
    "lapply(nog4_combined_const_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_const_bootstraps.RData')\n",
    "\n",
    "# non-G4 stable-G bootstraps\n",
    "nog4_combined_stable_bootstraps <-  lapply(promoter_nog4_combined_stable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "save(nog4_combined_stable_bootstraps, file = 'RData/nog4_combined_stable_bootstraps.RData')\n",
    "lapply(nog4_combined_stable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_stable_bootstraps.RData')\n",
    "\n",
    "# non-G4 unstable-G bootstraps\n",
    "nog4_combined_unstable_bootstraps <-  lapply(promoter_nog4_combined_unstable, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_delta, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_unstable_bootstraps, file = 'RData/nog4_combined_unstable_bootstraps.RData')\n",
    "lapply(nog4_combined_unstable_bootstraps, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_unstable_bootstraps.RData')\n",
    "       \n",
    "# # g4 in PLOF genes bootstrap\n",
    "combined_g4_plof_gruns_bootstrap <-  lapply(promoter_combined_g4_plof_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_plof_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_plof_gruns_bootstrap, file = 'RData/combined_g4_plof_gruns_bootstraps.RData')\n",
    "lapply(combined_g4_plof_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load(\"RData/combined_g4_plof_gruns_bootstraps.RData\")\n",
    "\n",
    "# # nonG4 in PLOF genes bootstrap\n",
    "combined_npg4_plof_gruns_bootstrap <-  lapply(nog4_combined_plof, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_plof, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_npg4_plof_gruns_bootstrap, file = 'RData/combined_npg4_plof_gruns_bootstrap.RData')\n",
    "lapply(combined_npg4_plof_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_npg4_plof_gruns_bootstrap.RData')\n",
    "# #\n",
    "# G4 in disease gene bootstrap\n",
    "combined_g4_clinvar_signal_bootstrap <-  lapply(combined_g4_clinvar_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))\n",
    "#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_clinvar_signal_bootstrap, file = 'RData/combined_g4_clinvar_signal_bootstrap.RData')\n",
    "lapply(combined_g4_clinvar_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_clinvar_signal_bootstrap.RData')\n",
    "\n",
    "# # G4 in cgcs gene bootstrap\n",
    "combined_g4_cgcs_signal_bootstrap <-  lapply(combined_g4_cgcs_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_cgcs_signal_bootstrap, file = 'RData/combined_g4_cgcs_signal_bootstrap.RData')\n",
    "lapply(combined_g4_cgcs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_cgcs_signal_bootstrap.RData')\n",
    "\n",
    "# # G4 in disease gene bootstrap\n",
    "combined_g4_tsgs_signal_bootstrap <-  lapply(combined_g4_tsgs_signal_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_gruns, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(combined_g4_tsgs_signal_bootstrap, file = 'RData/combined_g4_tsgs_signal_bootstrap.RData')\n",
    "lapply(combined_g4_tsgs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_tsgs_signal_bootstrap.RData')\n",
    "\n",
    "# # nonG4 in disease gene bootstrap\n",
    "nog4_combined_clinvar_signal_bootstrap <-  lapply(nog4_combined_clinvar_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm= ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_clinvar_signal_bootstrap, file = 'RData/nog4_combined_clivar_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_clinvar_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_clivar_signal_bootstrap.RData')\n",
    "# #\n",
    "# # nonG4 in disease gene bootstrap\n",
    "nog4_combined_cgcs_signal_bootstrap <-  lapply(nog4_combined_cgcs_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_cgcs_signal_bootstrap, file = 'RData/nog4_combined_cgcs_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_cgcs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_cgcs_signal_bootstrap.RData')\n",
    "# #\n",
    "# # nonG4 in disease gene bootstrap\n",
    "nog4_combined_tsgs_signal_bootstrap <-  lapply(nog4_combined_tsgs_signal, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(nog4_combined_signal, function(x) generate_bootstrap(x, 10000, mu = mu, ps_model = ps_lm))\n",
    "save(nog4_combined_tsgs_signal_bootstrap, file = 'RData/nog4_combined_tsgs_signal_bootstrap.RData')\n",
    "lapply(nog4_combined_tsgs_signal_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/nog4_combined_tsgs_signal_bootstrap.RData')\n",
    "\n",
    "# # delta g4 in disease genes bootstrap\n",
    "combined_g4_clinvar_signal_delta_gruns_bootstrap <-  lapply(combined_g4_clinvar_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_clinvar_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_clinvar_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_clinvar_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_clinvar_signal_delta_gruns_bootstrap.RData')\n",
    "\n",
    "# # delta g4 in cgcs genes bootstrap\n",
    "combined_g4_cgcs_signal_delta_gruns_bootstrap <-  lapply(combined_g4_cgcs_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_cgcs_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_cgcs_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_cgcs_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_cgcs_signal_delta_gruns_bootstrap.RData')\n",
    "\n",
    "# # delta g4 in tsgs genes bootstrap\n",
    "combined_g4_tsgs_signal_delta_gruns_bootstrap <-  lapply(combined_g4_tsgs_signal_delta_gruns, function(x) bootstrap_maps2(x, mu = mu, ps_lm = ps_lm_genomes, ntimes = 10000))#lapply(combined_g4_signal_delta_gruns,\n",
    "save(combined_g4_tsgs_signal_delta_gruns_bootstrap, file = 'RData/combined_g4_tsgs_signal_delta_gruns_bootstrap.RData')\n",
    "lapply(combined_g4_tsgs_signal_delta_gruns_bootstrap, function(x) quantile(x, c(0.05, 0.95)))\n",
    "# load('RData/combined_g4_tsgs_signal_delta_gruns_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants bootstrap\n",
    "promoter_gnomad_grange_bootstrap <- chunked_bootstrap(promoter_gnomad_grange, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(promoter_gnomad_grange_bootstrap, file = 'RData/promoter_gnomad_grange_bootstrap.RData')\n",
    "quantile(promoter_gnomad_grange_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/promoter_gnomad_grange_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants delta pG4 bootstrap\n",
    "combined_promoter_delta_variants_bootstrap <- chunked_bootstrap(all_promoter_delta_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_delta_variants_bootstrap, file = 'RData/combined_promoter_delta_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_delta_variants_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/combined_promoter_delta_variants_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants const pG4 bootstrap\n",
    "combined_promoter_const_variants_bootstrap <- chunked_bootstrap(const_all_promoter_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_const_variants_bootstrap, file = 'RData/combined_promoter_const_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_const_variants_bootstrap, c(0.05, 0.95))\n",
    "# load(\"RData/combined_promoter_const_variants_bootstrap.RData\")\n",
    "\n",
    "# ALL promoter variants stable pG4 bootstrap\n",
    "combined_promoter_stable_variants_bootstrap <- chunked_bootstrap(all_promoter_stable_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_stable_variants_bootstrap, file = 'RData/combined_promoter_stable_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_stable_variants_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/combined_promoter_stable_variants_bootstrap.RData')\n",
    "\n",
    "# ALL promoter variants unstable pG4 bootstrap\n",
    "combined_promoter_unstable_variants_bootstrap <- chunked_bootstrap(all_promoter_unstable_variants, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(combined_promoter_unstable_variants_bootstrap, file = 'RData/combined_promoter_unstable_variants_bootstrap.RData')\n",
    "quantile(combined_promoter_unstable_variants_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/combined_promoter_unstable_variants_bootstrap.RData')\n",
    "\n",
    "# # All promoter variants plof bootstrap\n",
    "all_promoter_plof_bootstrap <- chunked_bootstrap(all_promoter_plof, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_plof_bootstrap, file = 'RData/all_promoter_plof_bootstrap.RData')\n",
    "quantile(all_promoter_plof_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_plof_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_clinvar_dx_bootstrap <- chunked_bootstrap(all_promoter_clinvar_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_clinvar_dx_bootstrap, file = 'RData/all_promoter_clinvar_dx_bootstrap.RData')\n",
    "quantile(all_promoter_clinvar_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_clinvar_dx_bootstrap.RData')\n",
    "#\n",
    "# # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_cgcs_dx_bootstrap <- chunked_bootstrap(all_promoter_cgcs_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_cgcs_dx_bootstrap, file = 'RData/all_promoter_cgcs_dx_bootstrap.RData')\n",
    "quantile(all_promoter_cgcs_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_cgcs_dx_bootstrap.RData')\n",
    "\n",
    "# # ALL promoter variants disease genes bootstrap\n",
    "all_promoter_tsgs_dx_bootstrap <- chunked_bootstrap(all_promoter_tsgs_signal, 10000, mu = mu, ps_model = ps_lm_genomes)\n",
    "save(all_promoter_tsgs_dx_bootstrap, file = 'RData/all_promoter_tsgs_dx_bootstrap.RData')\n",
    "quantile(all_promoter_tsgs_dx_bootstrap, c(0.05, 0.95))\n",
    "# load('RData/all_promoter_tsgs_dx_bootstrap.RData')\n",
    "                                      \n",
    "confidence_intervals <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                             'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                             'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                             'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                             'Top-third SNV' = all_promoter_plof_bootstrap,\n",
    "                             'Top-third pG4 G1' = combined_g4_plof_gruns_bootstrap$G1,\n",
    "                             'Top-third pG4 G2' = combined_g4_plof_gruns_bootstrap$G2,\n",
    "                             'Top-third pG4 G3' = combined_g4_plof_gruns_bootstrap$G3,\n",
    "                             'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                             'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                             'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                             'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                             'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                             'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                             'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                             'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "\n",
    "                             'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                             'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                             'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                             'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3\n",
    "                             )\n",
    "\n",
    "confidence_intervals <- sapply(confidence_intervals, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                               maps_pg4_combined$G1['MAPS'],\n",
    "                               maps_pg4_combined$G2['MAPS'],\n",
    "                               maps_pg4_combined$G3['MAPS'],\n",
    "                               maps_plof_all_variants$all_promoter_plof['MAPS'],\n",
    "                               maps_plof_gruns$G1['MAPS'],\n",
    "                               maps_plof_gruns$G2['MAPS'],\n",
    "                               maps_plof_gruns$G3['MAPS'],\n",
    "                               maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                               maps_clinvar_pg4$G1['MAPS'],\n",
    "                               maps_clinvar_pg4$G2['MAPS'],\n",
    "                               maps_clinvar_pg4$G3['MAPS'],\n",
    "                               maps_cgcs_signal_all_variants$cgcs[\"MAPS\"],\n",
    "                               maps_cgcs_pg4$G1['MAPS'],\n",
    "                               maps_cgcs_pg4$G2[\"MAPS\"],\n",
    "                               maps_cgcs_pg4$G3[\"MAPS\"],\n",
    "                               maps_tsgs_signal_all_variants$tsgs[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G1[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G2[\"MAPS\"],\n",
    "                               maps_tsgs_pg4$G3[\"MAPS\"]\n",
    "                               )\n",
    "\n",
    "plot_df <- confidence_intervals\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"LOEUF Top 1/3\", \"LOEUF Top 1/3\", \"LOEUF Top 1/3\", \"LOEUF Top 1/3\",\n",
    "                             \"ClinVar Genes\", \"ClinVar Genes\", \"ClinVar Genes\", \"ClinVar Genes\",\n",
    "                             \"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\n",
    "                             \"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_plof),\n",
    "               sapply(combined_g4_plof_gruns, function(x) length(x)),\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_gruns, function(x) length(x))\n",
    ")\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "\n",
    "# tiff('figures/MAPS.tiff', units = 'in', width = 9, height = 6, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "#  geom_segment(x = 10, xend = 10, y = plot_df['ClinVar pG4 G1',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "#  geom_segment(x = 8, xend = 8, y = plot_df['Top-third pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, 0.10)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'red') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.4, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.0, y = 0.096, label = \"pLOF\", color = 'red', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'LOEUF Top 1/3', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 3)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 3)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 3)\n",
    "ggsave('figures/MAPS.tiff', units = 'in', width = 9, height = 6, device=tiff())\n",
    "ggsave('figures/MAPS.pdf', width = 9, height = 6,device=pdf())\n",
    "dev.off()\n",
    "\n",
    "###################################\n",
    "## Plot delta pG4 vs. const. pG4 ##\n",
    "###################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                               'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                               'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                               'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                               'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                               'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                               'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                               'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "                               'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                               'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                               'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                               'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3\n",
    "                             )\n",
    "\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "                                 maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                                 maps_clinvar_pg4$G1['MAPS'],\n",
    "                                 maps_clinvar_pg4$G2['MAPS'],\n",
    "                                 maps_clinvar_pg4$G3['MAPS'],\n",
    "                                 maps_cgcs_signal_all_variants$cgcs[\"MAPS\"],\n",
    "                                 maps_cgcs_pg4$G1['MAPS'],\n",
    "                                 maps_cgcs_pg4$G2[\"MAPS\"],\n",
    "                                 maps_cgcs_pg4$G3[\"MAPS\"],\n",
    "                                 maps_tsgs_signal_all_variants$tsgs[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G1[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G2[\"MAPS\"],\n",
    "                                 maps_tsgs_pg4$G3[\"MAPS\"]\n",
    "                               )\n",
    "\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"Alternative\", \"Alternative\", \"Alternative\", \"Alternative\",\n",
    "                             \"ClinVar\", \"ClinVar\", \"ClinVar\", \"ClinVar\",\n",
    "                             \"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\"Cancer Gene Census\",\n",
    "                             \"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\",\"Tumor suppressor gene\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_gruns, function(x) length(x)),\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_gruns, function(x) length(x))\n",
    ")\n",
    "\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "\n",
    "# tiff('figures/MAPS_delta_dx.tiff', units = 'in', width = 9, height = 6, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.4) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 0.8, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Diseae Genes', size = 3)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 3)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 3)\n",
    "\n",
    "ggsave('figures/MAPS_delta_dx.tiff', units = 'in', width = 9, height = 6,device=tiff())\n",
    "ggsave('figures/MAPS_delta_dx.pdf', width = 9, height = 6, device=pdf())\n",
    "dev.off()\n",
    "\n",
    "tiff('figures/MAPS_delta_dx_resized.tiff', units = 'in', width = 6, height = 4, res = 600)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 4),\n",
    "        axis.text.y = element_text(hjust = 1, size = 7),\n",
    "        axis.title.y = element_text(size=5),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.title = element_text(size=5),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 1.3, y = 0.043, label = \"missense\", color = 'orange', size = 1.5) +\n",
    "  annotate(\"text\", x = 1.55, y = 0.005, label = \"synonymous\", color = 'grey', size = 1.5) +\n",
    "  annotate(\"text\", x = 1, y = 0.096, label = \"pLOF\", color = 'purple', size = 1.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 2) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 2)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 2)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor suppressor gene', size = 2)\n",
    "dev.off()\n",
    "\n",
    "# pdf('figures/MAPS_delta_dx_resized.pdf', width = 6, height = 4, res = 600)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 12, xend = 12, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 5),\n",
    "        axis.text.y = element_text(hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(size=9),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.title = element_text(size=5),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 1.3, y = 0.043, label = \"missense\", color = 'orange', size = 1.5) +\n",
    "  annotate(\"text\", x = 1.55, y = 0.005, label = \"synonymous\", color = 'grey', size = 1.5) +\n",
    "  annotate(\"text\", x = 1, y = 0.096, label = \"pLOF\", color = 'purple', size = 1.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All Promoters', size = 2) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Disease Genes', size = 2)+\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Cancer Gene Census', size = 2)+\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Tumor Suppressor Gene', size = 2)\n",
    "# grDevices::embedFonts('figures/MAPS_delta_dx_resized.pdf')\n",
    "ggsave('figures/MAPS_delta_dx_resized.pdf', width = 6, height = 4,device = pdf())\n",
    "ggsave('figures/MAPS_delta_dx_resized.tiff', width = 6, height = 4,device = tiff())\n",
    "dev.off()\n",
    "\n",
    "###########################################\n",
    "## Plot all G4, delta G4, and clinvar G4 ##\n",
    "###########################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'Const. SNV' = combined_promoter_const_variants_bootstrap,\n",
    "                               'Const. pG4 G1' = g4_combined_const_bootstraps$G1,\n",
    "                               'Const. pG4 G2' = g4_combined_const_bootstraps$G2,\n",
    "                               'Const. pG4 G3' = g4_combined_const_bootstraps$G3,\n",
    "\n",
    "                               'Stable SNV' = combined_promoter_stable_variants_bootstrap,\n",
    "                               'Stable pG4 G1' = g4_combined_stable_bootstraps$G1,\n",
    "                               'Stable pG4 G2' = g4_combined_stable_bootstraps$G2,\n",
    "                               'Stable pG4 G3' = g4_combined_stable_bootstraps$G3,\n",
    "\n",
    "                               'Unstable SNV' = combined_promoter_unstable_variants_bootstrap,\n",
    "                               'Unstable pG4 G1' = g4_combined_unstable_bootstraps$G1,\n",
    "                               'Unstable pG4 G2' = g4_combined_unstable_bootstraps$G2,\n",
    "                               'Unstable pG4 G3' = g4_combined_unstable_bootstraps$G3\n",
    "\n",
    ")\n",
    "\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_const$all['MAPS'],\n",
    "                                 maps_promoter_combined_const$G1['MAPS'],\n",
    "                                 maps_promoter_combined_const$G2['MAPS'],\n",
    "                                 maps_promoter_combined_const$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_stable$all['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_unstable$all['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G3['MAPS']\n",
    "\n",
    "                                 )\n",
    "\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(\"All\", \"All\", \"All\", \"All\",\n",
    "                             \"Alternative\", \"Alternative\", \"Alternative\", \"Alternative\",\n",
    "                             \"Const.\", \"Const.\", \"Const.\", \"Const.\",\n",
    "                             \"Stable.\", \"Stable.\", \"Stable.\", \"Stable.\",\n",
    "                             \"Unstable.\", \"Unstable.\", \"Unstable.\", \"Unstable.\"\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               length(const_all_promoter_variants),\n",
    "               sapply(promoter_combined_const, function(x) length(x)),\n",
    "               length(all_promoter_stable_variants),\n",
    "               sapply(promoter_combined_stable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_unstable_variants),\n",
    "               sapply(promoter_combined_unstable, function(x) length(x))\n",
    "               )\n",
    "\n",
    "plot_df$Context <- paste0(rep(c('All SNV','G4 G1', 'G4 G2', 'G4 G3'), 5),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4'),5)\n",
    "# tiff('figures/MAPS_delta_const.tiff', units = 'in', width = 7, height = 5, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0)) +\n",
    "  geom_point() +\n",
    "  ylim(c(-0.05, .12)) + ylab(expression(atop(\"Increasing Negative Selection  \" ~ symbol('\\256'),\"MAPS\"))) +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\")) +\n",
    "  theme(axis.line = element_line(size = 1, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),\n",
    "        axis.title.y = element_text(),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\")) +\n",
    "  annotate(\"text\", x = 1, y = 0.044, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 0.8, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 4.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 8.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 12.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 16.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 2.5, y = -0.05, label = 'All UTRs', size = 3) +\n",
    "  annotate(\"text\", x = 6.5, y = -0.05, label = 'Alt. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 10.5, y = -0.05, label = 'Cons. pG4', size = 3) +\n",
    "  annotate(\"text\", x = 14.5, y = -0.05, label = 'Stable. UTRs', size = 3) +\n",
    "  annotate(\"text\", x = 18.5, y = -0.05, label = 'Unstable. pG4', size = 3)\n",
    "ggsave('figures/MAPS_delta_const.tiff', units = 'in', width = 12, height = 5, dpi = 300,device=tiff())\n",
    "ggsave('figures/MAPS_delta_const.pdf', width = 12, height = 5, device=pdf())\n",
    "dev.off()\n",
    "\n",
    "\n",
    "###########################################\n",
    "## Plot with comparison of non-pG4 Gruns ##\n",
    "###########################################\n",
    "\n",
    "confidence_intervals_2 <- list('All SNV' = promoter_gnomad_grange_bootstrap,\n",
    "                               'pG4 G1' = g4_combined_bootstraps$G1,\n",
    "                               'pG4 G2' = g4_combined_bootstraps$G2,\n",
    "                               'pG4 G3' = g4_combined_bootstraps$G3,\n",
    "                               'non-G4 1' = nog4_combined_bootstraps$pos1,\n",
    "                               'non-G4 2' = nog4_combined_bootstraps$pos2,\n",
    "                               'non-G4 3' = nog4_combined_bootstraps$pos3,\n",
    "                               'plof All SNV' = all_promoter_plof_bootstrap,\n",
    "                               'plof pG4 G1' = combined_g4_plof_gruns_bootstrap$G1,\n",
    "                               'plof pG4 G2' = combined_g4_plof_gruns_bootstrap$G2,\n",
    "                               'plof pG4 G3' = combined_g4_plof_gruns_bootstrap$G3,\n",
    "                               'plof non-G4 1' = combined_npg4_plof_gruns_bootstrap$pos1,\n",
    "                               'plof non-G4 2' = combined_npg4_plof_gruns_bootstrap$pos2,\n",
    "                               'plof non-G4 3' = combined_npg4_plof_gruns_bootstrap$pos3,\n",
    "                               'Delta SNV' = combined_promoter_delta_variants_bootstrap,\n",
    "                               'Delta pG4 G1' = g4_combined_delta_bootstraps$G1,\n",
    "                               'Delta pG4 G2' = g4_combined_delta_bootstraps$G2,\n",
    "                               'Delta pG4 G3' = g4_combined_delta_bootstraps$G3,\n",
    "                               'Delta non-G4 1' = nog4_combined_delta_bootstraps$pos1,\n",
    "                               'Delta non-G4 2' = nog4_combined_delta_bootstraps$pos2,\n",
    "                               'Delta non-G4 3' = nog4_combined_delta_bootstraps$pos3,\n",
    "                               'Const. SNV' = combined_promoter_const_variants_bootstrap,\n",
    "                               'Const. pG4 G1' = g4_combined_const_bootstraps$G1,\n",
    "                               'Const. pG4 G2' = g4_combined_const_bootstraps$G2,\n",
    "                               'Const. pG4 G3' = g4_combined_const_bootstraps$G3,\n",
    "                               'Const. non-G4 1' = nog4_combined_const_bootstraps$pos1,\n",
    "                               'Const. non-G4 2' = nog4_combined_const_bootstraps$pos2,\n",
    "                               'Const. non-G4 3' = nog4_combined_const_bootstraps$pos3,\n",
    "\n",
    "                               'Stable SNV' = combined_promoter_stable_variants_bootstrap,\n",
    "                               'Stable pG4 G1' = g4_combined_stable_bootstraps$G1,\n",
    "                               'Stable pG4 G2' = g4_combined_stable_bootstraps$G2,\n",
    "                               'Stable pG4 G3' = g4_combined_stable_bootstraps$G3,\n",
    "                               'Stable non-G4 1' = nog4_combined_stable_bootstraps$pos1,\n",
    "                               'Stable non-G4 2' = nog4_combined_stable_bootstraps$pos2,\n",
    "                               'Stable non-G4 3' = nog4_combined_stable_bootstraps$pos3,\n",
    "\n",
    "\n",
    "                               'Unstable SNV' = combined_promoter_unstable_variants_bootstrap,\n",
    "                               'Unstable pG4 G1' = g4_combined_unstable_bootstraps$G1,\n",
    "                               'Unstable pG4 G2' = g4_combined_unstable_bootstraps$G2,\n",
    "                               'Unstable pG4 G3' = g4_combined_unstable_bootstraps$G3,\n",
    "                               'Unstable non-G4 1' = nog4_combined_unstable_bootstraps$pos1,\n",
    "                               'Unstable non-G4 2' = nog4_combined_unstable_bootstraps$pos2,\n",
    "                               'Unstable non-G4 3' = nog4_combined_unstable_bootstraps$pos3,\n",
    "\n",
    "                               'ClinVar SNV' = all_promoter_clinvar_dx_bootstrap,\n",
    "                               'ClinVar pG4 G1' = combined_g4_clinvar_signal_bootstrap$G1,\n",
    "                               'ClinVar pG4 G2' = combined_g4_clinvar_signal_bootstrap$G2,\n",
    "                               'ClinVar pG4 G3' = combined_g4_clinvar_signal_bootstrap$G3,\n",
    "                               'ClinVar non-G4 1' = nog4_combined_clinvar_signal_bootstrap$pos1,\n",
    "                               'ClinVar non-G4 2' = nog4_combined_clinvar_signal_bootstrap$pos2,\n",
    "                               'ClinVar non-G4 3' = nog4_combined_clinvar_signal_bootstrap$pos3,\n",
    "\n",
    "                               'Cosmic SNV' = all_promoter_cgcs_dx_bootstrap,\n",
    "                               'Cosmic pG4 G1' = combined_g4_cgcs_signal_bootstrap$G1,\n",
    "                               'Cosmic pG4 G2' = combined_g4_cgcs_signal_bootstrap$G2,\n",
    "                               'Cosmic pG4 G3' = combined_g4_cgcs_signal_bootstrap$G3,\n",
    "                               'Cosmic non-G4 1' = nog4_combined_cgcs_signal_bootstrap$pos1,\n",
    "                               'Cosmic non-G4 2' = nog4_combined_cgcs_signal_bootstrap$pos2,\n",
    "                               'Cosmic non-G4 3' = nog4_combined_cgcs_signal_bootstrap$pos3,\n",
    "\n",
    "                               'TsGenes SNV' = all_promoter_tsgs_dx_bootstrap,\n",
    "                               'TsGenes pG4 G1' = combined_g4_tsgs_signal_bootstrap$G1,\n",
    "                               'TsGenes pG4 G2' = combined_g4_tsgs_signal_bootstrap$G2,\n",
    "                               'TsGenes pG4 G3' = combined_g4_tsgs_signal_bootstrap$G3,\n",
    "                               'TsGenes non-G4 1' = nog4_combined_tsgs_signal_bootstrap$pos1,\n",
    "                               'TsGenes non-G4 2' = nog4_combined_tsgs_signal_bootstrap$pos2,\n",
    "                               'TsGenes non-G4 3' = nog4_combined_tsgs_signal_bootstrap$pos3\n",
    ")\n",
    "confidence_intervals_2 <- sapply(confidence_intervals_2, function(x) quantile(x, c(0.05, 0.95))) %>%\n",
    "  t() %>%\n",
    "  as.data.frame()\n",
    "\n",
    "confidence_intervals_2$MAPS <- c(maps_all_promoter$all['MAPS'],\n",
    "                                 maps_pg4_combined$G1['MAPS'],\n",
    "                                 maps_pg4_combined$G2['MAPS'],\n",
    "                                 maps_pg4_combined$G3['MAPS'],\n",
    "                                 maps_nong4_first_last$pos1['MAPS'],\n",
    "                                 maps_nong4_first_last$pos2['MAPS'],\n",
    "                                 maps_nong4_first_last$pos3['MAPS'],\n",
    "\n",
    "                                 maps_plof_all_variants$all_promoter_plof['MAPS'],\n",
    "                                 maps_plof_gruns$G1['MAPS'],\n",
    "                                 maps_plof_gruns$G2['MAPS'],\n",
    "                                 maps_plof_gruns$G3['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos1['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos2['MAPS'],\n",
    "                                 maps_nog4_combined_plof$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_delta$all['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G1['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G2['MAPS'],\n",
    "                                 maps_promoter_combined_delta$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_delta$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_const$all['MAPS'],\n",
    "                                 maps_promoter_combined_const$G1['MAPS'],\n",
    "                                 maps_promoter_combined_const$G2['MAPS'],\n",
    "                                 maps_promoter_combined_const$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_const$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_stable$all['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_stable$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_stable$pos3['MAPS'],\n",
    "\n",
    "                                 maps_all_promoter_unstable$all['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G1['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G2['MAPS'],\n",
    "                                 maps_promoter_combined_unstable$G3['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos1['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos2['MAPS'],\n",
    "                                 maps_promoter_nog4_unstable$pos3['MAPS'],\n",
    "\n",
    "                                 maps_clinvar_signal_all_variants$clinvar['MAPS'],\n",
    "                                 maps_clinvar_pg4$G1['MAPS'],\n",
    "                                 maps_clinvar_pg4$G2['MAPS'],\n",
    "                                 maps_clinvar_pg4$G3['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos1['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos2['MAPS'],\n",
    "                                 maps_clinvar_nong4$pos3['MAPS'],\n",
    "\n",
    "                                 maps_cgcs_signal_all_variants$cgcs['MAPS'],\n",
    "                                 maps_cgcs_pg4$G1['MAPS'],\n",
    "                                 maps_cgcs_pg4$G2['MAPS'],\n",
    "                                 maps_cgcs_pg4$G3['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos1['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos2['MAPS'],\n",
    "                                 maps_cgcs_nong4$pos3['MAPS'],\n",
    "\n",
    "                                 maps_tsgs_signal_all_variants$tsgs['MAPS'],\n",
    "                                 maps_tsgs_pg4$G1['MAPS'],\n",
    "                                 maps_tsgs_pg4$G2['MAPS'],\n",
    "                                 maps_tsgs_pg4$G3['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos1['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos2['MAPS'],\n",
    "                                 maps_tsgs_nong4$pos3['MAPS']\n",
    ")\n",
    "plot_df <- confidence_intervals_2\n",
    "plot_df$Category <- factor(c(rep(\"All\", 7),\n",
    "                             rep('Top 1/3 LOEUF', 7),\n",
    "                             rep(\"Alternative\", 7),\n",
    "                             rep('Constitutive', 7),\n",
    "                             rep(\"Stable\", 7),\n",
    "                             rep('Unstable', 7),\n",
    "                             rep('ClinVar', 7),\n",
    "                             rep('Cancer Gene Cencus',7),\n",
    "                             rep('Tumor Suppressor Genes',7)\n",
    "                             ),\n",
    "                           ordered = TRUE)\n",
    "plot_df$N <- c(length(promoter_gnomad_grange),\n",
    "               sapply(promoter_combined, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_plof),\n",
    "               sapply(combined_g4_plof_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_plof, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_delta_variants),\n",
    "               sapply(promoter_combined_delta, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_delta, function(x) length(x)),\n",
    "\n",
    "               length(const_all_promoter_variants),\n",
    "               sapply(promoter_combined_const, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_const, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_stable_variants),\n",
    "               sapply(promoter_combined_stable, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_stable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_unstable_variants),\n",
    "               sapply(promoter_combined_unstable, function(x) length(x)),\n",
    "               sapply(promoter_nog4_combined_unstable, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_clinvar_signal),\n",
    "               sapply(combined_g4_clinvar_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_clinvar_signal, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_cgcs_signal),\n",
    "               sapply(combined_g4_cgcs_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_cgcs_signal, function(x) length(x)),\n",
    "\n",
    "               length(all_promoter_tsgs_signal),\n",
    "               sapply(combined_g4_tsgs_signal_delta_gruns, function(x) length(x)),\n",
    "               sapply(nog4_combined_tsgs_signal, function(x) length(x))\n",
    "               )\n",
    "plot_df$Context <- paste0(rep(c('All SNV','pG4 G1', 'pG4 G2', 'pG4 G3', 'non-G4 1', 'non-G4 2', 'non-G4 3'), 9),\n",
    "                          '\\n(N = ',\n",
    "                          prettyNum(plot_df$N, big.mark=',', trim = TRUE), ')')\n",
    "\n",
    "plot_df$Context <- factor(plot_df$Context, levels = plot_df$Context)\n",
    "\n",
    "plot_df$Class <- rep(c('All SNVs', 'G4', 'G4', 'G4', 'non-G4', 'non-G4', 'non-G4'),9)\n",
    "\n",
    "# tiff('figures/MAPS_all_classes_3.tiff', units = 'in', width = 10, height = 4, res = 300)\n",
    "ggplot(plot_df, aes(x = Context, y = MAPS, color = Class)) +\n",
    "  geom_errorbar(aes(ymin=`5%`, ymax=`95%`, width=0), size = 0.25) +\n",
    "  # geom_segment(x = 32, xend = 32, y = plot_df['ClinVar pG4 G3',]$`95%`, yend = -Inf, color = 'red', size = 0.2) +\n",
    "  geom_point(size = 0.5) +\n",
    "  ylim(c(-0.05, .12)) + ylab(\"MAPS\") +\n",
    "  xlab('') +\n",
    "  geom_hline(yintercept = 0, lty = 2, color = 'grey') +\n",
    "  geom_hline(yintercept = maps_cds_markers$missense_all['MAPS'], lty = 2, color = 'orange') +\n",
    "  geom_hline(yintercept = maps_cds_markers$lof_all['MAPS'], lty = 2, color = 'purple') +\n",
    "  scale_color_manual(values = c(\"black\", \"red\", \"blue\")) +\n",
    "  theme(axis.line = element_line(size = 0.5, color = 'black'),\n",
    "        panel.background = element_rect((fill = 'white')),\n",
    "        axis.text.x = element_text(angle = 65, hjust = 1, size = 5),\n",
    "        axis.title.y = element_text(size = 8),\n",
    "        axis.text.y = element_text(size = 8),\n",
    "        plot.margin=unit(c(0,0,0,0),\"cm\"),\n",
    "        legend.position = 'top',\n",
    "        legend.title = element_blank(),\n",
    "        legend.text = element_text(size=5)) +\n",
    "  annotate(\"text\", x = 2, y = 0.043, label = \"missense\", color = 'orange', size = 2.5) +\n",
    "  annotate(\"text\", x = 2, y = 0.005, label = \"synonymous\", color = 'grey', size = 2.5) +\n",
    "  annotate(\"text\", x = 1.5, y = 0.096, label = \"pLOF\", color = 'purple', size = 2.5) +\n",
    "  geom_vline(xintercept = 7.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 14.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 15 + 6.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 28.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 35.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 42.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 49.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  geom_vline(xintercept = 56.5, lty = 1, color = 'darkgrey', size = 0.3) +\n",
    "  annotate(\"text\", x = 3.75, y = -0.05, label = 'All Promiters', size = 2) +\n",
    "  annotate(\"text\", x = 11.25, y = -0.05, label = 'Top 1/3 LOEUF', size = 2) +\n",
    "  annotate(\"text\", x = 17.75, y = -0.05, label = 'Alt. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 25.25, y = -0.05, label = 'Const. pG4', size = 2) +\n",
    "  annotate(\"text\", x = 24.25 + 7.5, y = -0.05, label = 'Stable pG4', size = 2) +\n",
    "  annotate(\"text\", x = 32.25 + 7, y = -0.05, label = 'Unstable pG4', size = 2) +\n",
    "  annotate(\"text\", x = 39.25 + 7.5, y = -0.05, label = 'ClinVar pG4', size = 2)+\n",
    "  annotate(\"text\", x = 46.25 + 7, y = -0.05, label = 'Cancer Gene Cencus pG4', size = 2)+\n",
    "  annotate(\"text\", x = 54.25 + 7, y = -0.05, label = 'Tumor Suppressor Genes pG4', size = 2)\n",
    "ggsave('figures/MAPS_all_classes_3.tiff', units = 'in', width = 10, height = 4, dpi = 300,device = tiff())\n",
    "ggsave('figures/MAPS_all_classes_3.pdf', width = 10, height = 4, device = pdf())\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b17e6ca",
   "metadata": {},
   "source": [
    "## R PopGenome analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efae3bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "# @Time    : 2022/11/8 9:30\n",
    "# @Author  : Guangyuel\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import pyranges as pr\n",
    "\n",
    "\n",
    "def file_check(path):\n",
    "    if not os.path.isfile(path):\n",
    "        sys,exit(f\"this file {path} is not exist !!!\")\n",
    "    else:\n",
    "        return path\n",
    "\n",
    "\n",
    "def promoter_select(promoter, promoter_pG4, promoter_name, gruns, WD_PATH):\n",
    "    promoter_pG4 = promoter_pG4.overlap(promoter, strandedness=False)\n",
    "    promoter_pG4_Stable = promoter_pG4[promoter_pG4.Score > 19]\n",
    "    promoter_pG4_unStable = promoter_pG4[promoter_pG4.Score <= 19]\n",
    "    promoter_noG4 = promoter.subtract(promoter_pG4, strandedness=False).merge()\n",
    "    promoter_pG4.to_csv(os.path.join(WD_PATH, promoter_name + \"_pG4.info.tsv\"), sep=\"\\t\")\n",
    "    promoter_pG4_Stable.to_csv(os.path.join(WD_PATH, promoter_name + \"_pG4_Stable.info.tsv\"), sep=\"\\t\")\n",
    "    promoter_pG4_unStable.to_csv(os.path.join(WD_PATH, promoter_name + \"_pG4_unStable.info.tsv\"), sep=\"\\t\")\n",
    "    promoter_noG4.to_csv(os.path.join(WD_PATH, promoter_name + \"_noG4.info.tsv\"), sep=\"\\t\")\n",
    "    # G-tracts analysis\n",
    "    promoter_G4_Gtract = gruns.overlap(promoter_pG4, strandedness=\"same\")\n",
    "    promoter_stableG4_Gtract = gruns.overlap(promoter_pG4_Stable, strandedness=\"same\")\n",
    "    promoter_unstable_Gtracts = gruns.overlap(promoter_pG4_unStable, strandedness=\"same\")\n",
    "    promoter_nonG4_Gtract = gruns.overlap(promoter_noG4, strandedness=False)\n",
    "    promoter_G4_Gtract.to_bed(os.path.join(WD_PATH,promoter_name + \"_pG4_Gtract.bed\"))\n",
    "    promoter_stableG4_Gtract.to_bed(os.path.join(WD_PATH, promoter_name + \"_stableG4_Gtract.bed\"))\n",
    "    promoter_unstable_Gtracts.to_bed(os.path.join(WD_PATH, promoter_name + \"_unstable_Gtracts.bed\"))\n",
    "    promoter_nonG4_Gtract.to_bed(os.path.join(WD_PATH,promoter_name + \"_nonG4_Gtract.bed\"))\n",
    "\n",
    "\n",
    "WD_PATH = \"/disk2/G4_MS/frequency/envolution/data\"\n",
    "DATA_PTAH = \"/disk2/G4_MS/promoter_select/promoter_info\"\n",
    "REFERENCE_PTAH = \"/disk2/G4_MS/reference\"\n",
    "\n",
    "promoter_pG4_path = file_check(os.path.join(DATA_PTAH, \"promoter_pG4.info.tsv\"))\n",
    "promoter_pG4 = pr.PyRanges(pd.read_table(promoter_pG4_path))\n",
    "promoter_path = file_check(os.path.join(DATA_PTAH, \"promoter.annotion.tsv\"))\n",
    "promoter = pr.PyRanges(pd.read_table(promoter_path))\n",
    "\n",
    "cds_path = file_check(os.path.join(REFERENCE_PTAH, \"gencode38.cds_region.bed\"))\n",
    "cds = pr.read_bed(cds_path)\n",
    "\n",
    "gruns_path = file_check(os.path.join(REFERENCE_PTAH, \"genomic_Gruns.bed\"))\n",
    "gruns = pr.PyRanges(pd.read_table(gruns_path, header=None, names=[\"Chromosome\", \"Start\", \"End\", \"Strand\", \"Gtract\"]))\n",
    "# promoter_pG4 = promoter_pG4.overlap(cds, strandedness=False, invert=True)\n",
    "promoter = promoter.subtract(cds, strandedness=False).max_disjoint(strand=False)\n",
    "promoter.to_bed(os.path.join(WD_PATH, \"promoter.bed\"))\n",
    "promoter_select(promoter, promoter_pG4, \"promoter\", gruns, WD_PATH)\n",
    "## activate promoter\n",
    "active_promoter_path = file_check(os.path.join(DATA_PTAH, \"promoter_info.annotion.H3K4me3.tsv\"))\n",
    "active_promoter = pr.PyRanges(pd.read_table(active_promoter_path))\n",
    "active_promoter_pG4_path = file_check(os.path.join(DATA_PTAH, \"promoter_pG4.info.H3K4me3.tsv\"))\n",
    "active_promoter_pG4 = pr.PyRanges(pd.read_table(active_promoter_pG4_path))\n",
    "active_promoter = active_promoter.subtract(cds, strandedness=False).max_disjoint(strand=False)\n",
    "active_promoter.to_bed(os.path.join(WD_PATH, \"active_promoter.bed\"))\n",
    "promoter_select(active_promoter, active_promoter_pG4, \"active_promoter\", gruns, WD_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5936767d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /disk2/G4_MS/frequency/envolution\n",
    "cat promoter.bed active_promoter.bed |cut -f1-3 |bedtools sort -i -|bedtools merge -i - > promoter.merge.bed\n",
    "awk -v OFS=\"\\t\" '{print$1,$2,$3-1}' promoter.merge.bed > promoter.merge.bed1\n",
    "mv promoter.merge.bed1 promoter.merge_bed\n",
    "for i in chr{{1..22},X,Y};do \n",
    "bcftools filter -R promoter.merge.bed -O z ../CCDG_13607_B01_GRM_WGS_2019-02-19_${i}.recalibrated_variants.norm_PASS_SNP.gz -o CCDG_13607_B01_GRM_WGS_2019-02-19_${i}.recalibrated_variants.norm_PASS_SNP.gz;done\n",
    "# bcftools concat  ../*recalibrated_variants.norm_PASS_SNP.gz -O z -o CCDG_13607_B01_GRM_WGS_2019-02-19_merge.recalibrated_variants.norm_PASS_SNP.gz\n",
    "bcftools concat  *recalibrated_variants.norm_PASS_SNP.gz -O z -R promoter.merge.bed -a -o CCDG_13607_B01_GRM_WGS_2019-02-19_merge.recalibrated_variants.norm_PASS_SNP.gz\n",
    "cut -f 1-3 promoter.bed |bedtools sort -i - |bedtools merge -i - | awk -v OFS=\"\\t\" '{print$1,$2,$3-1}'> promoter_bed\n",
    "cut -f 1-3 active_promoter.bed |bedtools sort -i - |bedtools merge -i - | awk -v OFS=\"\\t\" '{print$1,$2,$3-1}'> active_promoter_bed\n",
    "bcftools filter -R promoter.bed -O z CCDG_13607_B01_GRM_WGS_2019-02-19_merge.recalibrated_variants.norm_PASS_SNP.gz -o promoter.kg.norm_PASS_SNP.vcf.gz\n",
    "bcftools filter -R active_promoter.bed -O z CCDG_13607_B01_GRM_WGS_2019-02-19_merge.recalibrated_variants.norm_PASS_SNP.gz -o active_promoter.kg.norm_PASS_SNP.vcf.gz\n",
    "bcftools index -t active_promoter.kg.norm_PASS_SNP.vcf.gz \n",
    "bcftools index -t promoter.kg.norm_PASS_SNP.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fb7043",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /disk2/G4_MS/frequency/evolution/analysis\n",
    "tail -n +2 ../data/promoter_pG4.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > promoter_pG4.bed\n",
    "tail -n +2 ../data/promoter_pG4_Stable.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > promoter_pG4_Stable.bed\n",
    "tail -n +2 ../data/promoter_pG4_unStable.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > promoter_pG4_unStable.bed\n",
    "tail -n +2 ../data/active_promoter_pG4.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > active_promoter_pG4.bed\n",
    "tail -n +2 ../data/active_promoter_pG4_Stable.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > active_promoter_pG4_Stable.bed\n",
    "tail -n +2 ../data/active_promoter_pG4_unStable.info.tsv |cut -f 1-3|bedtools sort -i -|bedtools merge -i -|awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' > active_promoter_pG4_unStable.bed\n",
    "\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/promoter_pG4_Gtract.bed > promoter_pG4_Gtract.bed\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/promoter_stableG4_Gtract.bed > promoter_stableG4_Gtract.bed\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/promoter_unstable_Gtracts.bed > promoter_unstable_Gtracts.bed\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/active_promoter_pG4_Gtract.bed > active_promoter_pG4_Gtract.bed\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/active_promoter_stableG4_Gtract.bed > active_promoter_stableG4_Gtract.bed\n",
    "awk -v OFS=\"\\t\" '{print $1,$2-1,$3-1}' ../data/active_promoter_unstable_Gtracts.bed > active_promoter_unstable_Gtracts.bed\n",
    "mkdir pG4_popGenome\n",
    "bcftools filter -R promoter_pG4.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_popGenome/promoter_pG4.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R promoter_pG4_Stable.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_popGenome/promoter_pG4_Stable.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R promoter_pG4_unStable.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_popGenome/promoter_pG4_unStable.kg.norm_PASS_SNP.vcf\n",
    "mkdir pG4_Gtract_popGenome\n",
    "bcftools filter -R promoter_pG4_Gtract.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_Gtract_popGenome/promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R promoter_stableG4_Gtract.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_Gtract_popGenome/promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R promoter_unstable_Gtracts.bed ../data/promoter.kg.norm_PASS_SNP.vcf.gz -O v -o pG4_Gtract_popGenome/promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf\n",
    "\n",
    "mkdir eG4_popGenome\n",
    "bcftools filter -R active_promoter_pG4.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_popGenome/active_promoter_pG4.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R active_promoter_pG4_Stable.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_popGenome/active_promoter_pG4_Stable.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R active_promoter_pG4_unStable.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_popGenome/active_promoter_pG4_unStable.kg.norm_PASS_SNP.vcf\n",
    "mkdir eG4_Gtract_popGenome\n",
    "bcftools filter -R active_promoter_pG4_Gtract.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_Gtract_popGenome/active_promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R active_promoter_stableG4_Gtract.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_Gtract_popGenome/active_promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "bcftools filter -R active_promoter_unstable_Gtracts.bed ../data/active_promoter.kg.norm_PASS_SNP.vcf.gz -O v -o eG4_Gtract_popGenome/active_promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf\n",
    "\n",
    "grep -v \"chrX\" pG4_popGenome/promoter_pG4.kg.norm_PASS_SNP.vcf|grep -v \"chrY\" |grep -v \"chrM\" > pG4_popGenome/autosome.promoter_pG4.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" pG4_popGenome/promoter_pG4_Stable.kg.norm_PASS_SNP.vcf|grep -v \"chrY\" |grep -v \"chrM\" > pG4_popGenome/autosome.promoter_pG4_Stable.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" pG4_popGenome/promoter_pG4_unStable.kg.norm_PASS_SNP.vcf|grep -v \"chrY\" |grep -v \"chrM\" > pG4_popGenome/autosome.promoter_pG4_unStable.kg.norm_PASS_SNP.vcf\n",
    "\n",
    "grep -v \"chrX\" pG4_Gtract_popGenome/promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > pG4_Gtract_popGenome/autosome.promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" pG4_Gtract_popGenome/promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > pG4_Gtract_popGenome/autosome.promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" pG4_Gtract_popGenome/promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > pG4_Gtract_popGenome/autosome.promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf\n",
    "\n",
    "grep -v \"chrX\" eG4_popGenome/active_promoter_pG4.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > eG4_popGenome/autosome.active_promoter_pG4.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" eG4_popGenome/active_promoter_pG4_Stable.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > eG4_popGenome/autosome.active_promoter_pG4_Stable.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" eG4_popGenome/active_promoter_pG4_unStable.kg.norm_PASS_SNP.vcf |grep -v \"chrY\" |grep -v \"chrM\" > eG4_popGenome/autosome.active_promoter_pG4_unStable.kg.norm_PASS_SNP.vcf\n",
    "\n",
    "grep -v \"chrX\" eG4_Gtract_popGenome/active_promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf |grep -v \"chrY\"|grep -v \"chrM\" > eG4_Gtract_popGenome/autosome.active_promoter_pG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" eG4_Gtract_popGenome/active_promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf |grep -v \"chrY\"|grep -v \"chrM\" > eG4_Gtract_popGenome/autosome.active_promoter_stableG4_Gtract.kg.norm_PASS_SNP.vcf\n",
    "grep -v \"chrX\" eG4_Gtract_popGenome/active_promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf |grep -v \"chrY\"|grep -v \"chrM\" > eG4_Gtract_popGenome/autosome.active_promoter_unstable_Gtracts.kg.norm_PASS_SNP.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b9e86d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## work.R\n",
    "library(\"PopGenome\")\n",
    "args <- commandArgs(T)\n",
    "g4_type = args[1]\n",
    "vcfpath = args[2]\n",
    "setwd(\"/disk2/G4_MS/frequency/evolution/analysis\")\n",
    "\n",
    "#pG4 <- readData(\"pG4_Gtract_popGenome/promoter_stableG4_Gtract\", format=\"VCF\")\n",
    "pG4 <- readData(vcfpath, format=\"VCF\")\n",
    "\n",
    "neut <- neutrality.stats(pG4)\n",
    "df1 <- data.frame(Tajima.D=c(neut@Tajima.D),Fu.Li.D=c(neut@Fu.Li.D), Fu.Li.F=c(neut@Fu.Li.F))\n",
    "df1$groups = gsub(\"_\", \" \", g4_type)\n",
    "row.names(df1) = \"Merge\"\n",
    "population_select <- function(population){\n",
    "    new_population <- c()\n",
    "    for (i in population){\n",
    "        new_population <- c(new_population, i, paste(i,\".2\",sep=\"\") )\n",
    "    }\n",
    "    return(new_population)\n",
    "}\n",
    "\n",
    "AFR <- read.table(\"../data/population_AFR.csv\",header=F)\n",
    "EUR <- read.table(\"../data/population_EUR.csv\",header=F)\n",
    "AMR <- read.table(\"../data/population_AMR.csv\",header=F)\n",
    "EAS <- read.table(\"../data/population_EAS.csv\",header=F)\n",
    "SAS <- read.table(\"../data/population_SAS.csv\",header=F)\n",
    "\n",
    "AFR.2 = population_select(AFR)\n",
    "AMR.2 = population_select(AMR)\n",
    "SAS.2 = population_select(SAS)\n",
    "EAS.2 = population_select(EAS)\n",
    "EUR.2 = population_select(EUR)\n",
    "\n",
    "neut <- set.populations(neut, list(EUR=EUR.2, AFR=AFR.2,AMR=AMR.2,SAS=SAS.2,EAS=EAS.2))\n",
    "neut <- neutrality.stats(neut)\n",
    "df <- data.frame(Tajima.D=c(neut@Tajima.D), Fu.Li.D=c(neut@Fu.Li.D), Fu.Li.F=c(neut@Fu.Li.F))\n",
    "df$groups = gsub(\"_\",\" \", g4_type)\n",
    "row.names(df) <- c(\"EUR\", \"AFR\", \"AMR\", \"SAS\",\"EAS\")\n",
    "df <- rbind(df1,df)\n",
    "write.table(df, paste(g4_type,\".evolution.tsv\",sep=\"\"),sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5bc7970",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /disk2/G4_MS/frequency/evolution/analysis\n",
    "Rscript work.R active_promoter_pG4_Gtract eG4_Gtract_popGenome/active_promoter_pG4_Gtract\n",
    "Rscript work.R active_promoter_stableG4_Gtract eG4_Gtract_popGenome/active_promoter_stableG4_Gtract\n",
    "Rscript work.R active_promoter_unstable_Gtracts eG4_Gtract_popGenome/active_promoter_unstable_Gtracts\n",
    "Rscript work.R active_promoter_pG4 eG4_popGenome/active_promoter_pG4\n",
    "Rscript work.R active_promoter_pG4_Stable eG4_popGenome/active_promoter_pG4_Stable\n",
    "Rscript work.R active_promoter_pG4_unStable eG4_popGenome/active_promoter_pG4_unStable\n",
    "\n",
    "Rscript work.R promoter_pG4_Gtract pG4_Gtract_popGenome/promoter_pG4_Gtract\n",
    "Rscript work.R promoter_stableG4_Gtract pG4_Gtract_popGenome/promoter_stableG4_Gtract\n",
    "Rscript work.R promoter_unstable_Gtracts pG4_Gtract_popGenome/promoter_unstable_Gtracts\n",
    "\n",
    "Rscript work.R promoter_pG4 pG4_popGenome/promoter_pG4\n",
    "Rscript work.R promoter_pG4_Stable pG4_popGenome/promoter_pG4_Stable\n",
    "Rscript work.R promoter_pG4_unStable pG4_popGenome/promoter_pG4_unStable\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8b5f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# result filter and export\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def file_get(path):\n",
    "    file_list = []\n",
    "    files = os.listdir(path)\n",
    "    for f in files:\n",
    "        if f.endswith(\"evolution.tsv\"):\n",
    "            file_list.append(os.path.join(path,f))\n",
    "    return file_list\n",
    "            \n",
    "            \n",
    "wd = \"/disk2/G4_MS/frequency/evolution/analysis\"\n",
    "df_list = []\n",
    "for f in file_get(wd):\n",
    "    df_list.append(pd.read_table(f))\n",
    "df = pd.concat(df_list)\n",
    "df = df.iloc[:,[3,0,1,2]]\n",
    "df = df.round(3)\n",
    "\n",
    "with pd.ExcelWriter(os.path.join(wd,\"Evolution.xlsx\")) as writer:\n",
    "    df[df.index==\"Merge\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"1000g evolution\", index=False)\n",
    "    df[df.index==\"EUR\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"EUR\", index=False)\n",
    "    df[df.index==\"EAS\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"EAS\", index=False)\n",
    "    df[df.index==\"AFR\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"AFR\", index=False)\n",
    "    df[df.index==\"AMR\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"AMR\", index=False)\n",
    "    df[df.index==\"SAS\"].sort_values(\"Tajima.D\").to_excel(writer,sheet_name=\"SAS\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
